{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "8b4bd12f-ca3e-48ee-ad61-117a605640b3",
   "metadata": {},
   "source": [
    "### Clinvar variants"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2fe70563-cbf9-41db-b5fa-7529ca0e5da4",
   "metadata": {
    "tags": []
   },
   "source": [
    "Retrieval:\n",
    "1. Downloaded form ClinVar FTP website: https://ftp.ncbi.nlm.nih.gov/pub/clinvar/vcf_GRCh38/clinvar_20230527.vcf.gz\n",
    "   `/ibex/scratch/projects/c2102/databases/clinvar/2023_05/clinvar_20230527.vcf (decompressed)`\n",
    "\n",
    "2. Split into chromosomes with bcftools\n",
    "   `/ibex/scratch/projects/c2102/databases/clinvar/2023_05/clinvar_chr##.vcf`\n",
    "\n",
    "Processing:\n",
    "1. Annotated with VEP\n",
    "   `/ibex/scratch/projects/c2102/databases/clinvar/2023_05/chr#/chr#_x##.vcf.vep`  \n",
    "\n",
    "2. Filter out variants that contain `'not_specified'` or `'not_provided'` in the CLNDN field (phenotype/disease)\n",
    "\n",
    "3. Select variants that contain `'missense_variant'` in the `'Consequence'` column (from VEP)\n",
    "\n",
    "4. Select varinats that contain either `'Likely_pathogenic'` or `'Pathogenic'` in the `'CLNSIG'` column (from ClinVar)\n",
    "\n",
    "5. Select variants where the gene provided by ClinVar (`GENEINFO` column) is equal to the gene in the VEP annotation (`SYMBOL` column)\n",
    "\n",
    "6. Select variants where the RefSeq ID (RefSeq column) is equal to the transcript indicated in the `MANE_SELECT` or `MANE_PLUS_CLINICAL` columns (from VEP)\n",
    "\n",
    "7. Obtained UniProt IDs for each variant, from two sources (sometimes there's more than 1 UniProt ID for a given transcript):  \n",
    "   a) `UNIPROT_ISOFORM` column (from VEP)  \n",
    "   b) UniProt database (`/ibex/projects/c2102/databases/uniprot/2023_02/current_release/knowledgebase/idmapping/by_organism/HUMAN_9606_idmapping.dat.gz`)\n",
    "\n",
    "8. Get the AlphaFold model for each transcript, from two sources:  \n",
    "   a) `MANE` dataset from the AlphaFold Database (`/ibex/scratch/projects/c2102/databases/alphafold/alphafold_mane`)  \n",
    "   b) Regular AlphaFold Database (`/ibex/scratch/projects/c2102/databases/alphafold/alphafold_human_v4` )\n",
    "\n",
    "   Checked that the sequence length and residue match.\n",
    "\n",
    "9. Obtained intra-molecular contacts (`8 Angstroms` away, will change to 6)\n",
    "\n",
    "10. Obtained catalytic sites for residue of interest and residues 8 Angstroms away (`variant_annotator.structure_features.catalyticSites`)\n",
    "\n",
    "11. Annotated AlphaFold models with DSSP to calculate secondary structure and surface accessible area for every residue, and extracted the values for the variants.\n",
    "\n",
    "12. Extracted pLDDT scores in a window of 5 residues centered around the residue of interest (proxy for disorder)\n",
    "\n",
    "13. Annotated with FoldX  "
   ]
  },
  {
   "cell_type": "raw",
   "id": "d7a86467-f8fc-43b8-9dbe-30db7902fccd",
   "metadata": {
    "tags": []
   },
   "source": [
    "Potentially useful columns:  \n",
    "\n",
    "CLNDN                ClinVar's preferred disease name for the concept specified by disease identifiers in CLNDISDB  \n",
    "CLNSIG               clinical significance for this single variant  \n",
    "UniProt_IDs          list of UniProt IDs found for this transcript  \n",
    "PDB_path             path to the AF model  \n",
    "intra_contacts       list of intra-molecular contacts (currently 8 A away, will change to 6)  \n",
    "is_catalytic         if the residue of interest of any residue 8 A away is annotated as catalytic  \n",
    "pLDDT                mean pLDDT for 5-residue window centered in residue of interest  \n",
    "secondary_structre   secondary structure assignment by DSSP (see Output_short section in https://swift.cmbi.umcn.nl/gv/dssp/)  \n",
    "accessibility        residue water exposed surface in Angstrom^2, as annotated by DSSP  \n",
    "SIFT                 pathogenicity score (Stefan didn't like it)  \n",
    "PolyPhen             pathogenicity score (Stefan didn't like it)  \n",
    "Conservation         conservation score from the Ensembl Compara database. GERP scores from the EPO 35 way mammalian alignment. Positive scores represent highly-conserved positions while negative scores represent highly-variable positions.  \n",
    "BLOSUM62             score for conservative vs nonconservative amino acid changes   (https://www.labxchange.org/library/items/lb:LabXchange:24d0ec21:lx_image:1#:~:text=BLOSUM%20scores%20for%20substituting%20the,are%20classified%20as%20non%2Dconservative.)  \n",
    "DOMAINS              names of overlapping protein domains from different sources  \n",
    "last 22 columns from FoldX     https://foldxsuite.crg.eu/command/BuildModel  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "2a4ddf5e-8d13-4968-86eb-d3938bf4db7c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import sys\n",
    "import os\n",
    "import numpy as np\n",
    "from Bio.Data import IUPACData"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "17e316de-424b-4284-853c-f09ad4a635c7",
   "metadata": {},
   "outputs": [],
   "source": [
    "protein_letters_3to1 = {value:key for key,value in IUPACData.protein_letters_1to3.items()}\n",
    "protein_letters_1to3 = {key:value for key,value in IUPACData.protein_letters_1to3.items()}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "2bb3b513-5985-4c6d-ba4f-3620aa1761aa",
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.set_option('display.max_rows', 500)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "6f4aae7c-b638-4d50-a6b1-1cb660cbdc38",
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.set_option('display.min_rows', 15)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "4b6c2503-093a-4672-b837-6336311f23be",
   "metadata": {},
   "outputs": [],
   "source": [
    "clinvar_variants = pd.read_pickle('variants/clinvar_variants.pkl')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "277e2eab-efbb-4362-8116-f8d9a21c0e0d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(31252, 100)"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinvar_variants.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "8469547d-cc46-4a3c-aad4-c7aefa70f5a1",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['#CHROM',\n",
       " 'POS',\n",
       " 'ID',\n",
       " 'REF',\n",
       " 'ALT',\n",
       " 'AF_ESP',\n",
       " 'AF_EXAC',\n",
       " 'AF_TGP',\n",
       " 'ALLELEID',\n",
       " 'CLNDN',\n",
       " 'CLNDNINCL',\n",
       " 'CLNDISDB',\n",
       " 'CLNDISDBINCL',\n",
       " 'CLNHGVS',\n",
       " 'CLNREVSTAT',\n",
       " 'CLNSIG',\n",
       " 'CLNSIGCONF',\n",
       " 'CLNSIGINCL',\n",
       " 'CLNVC',\n",
       " 'CLNVCSO',\n",
       " 'CLNVI',\n",
       " 'DBVARID',\n",
       " 'MC',\n",
       " 'ORIGIN',\n",
       " 'RS',\n",
       " 'Allele',\n",
       " 'Consequence',\n",
       " 'IMPACT',\n",
       " 'SYMBOL',\n",
       " 'Gene',\n",
       " 'Feature_type',\n",
       " 'Feature',\n",
       " 'BIOTYPE',\n",
       " 'EXON',\n",
       " 'INTRON',\n",
       " 'HGVSc',\n",
       " 'HGVSp',\n",
       " 'cDNA_position',\n",
       " 'CDS_position',\n",
       " 'Protein_position',\n",
       " 'Amino_acids',\n",
       " 'Codons',\n",
       " 'Existing_variation',\n",
       " 'DISTANCE',\n",
       " 'STRAND',\n",
       " 'VARIANT_CLASS',\n",
       " 'SYMBOL_SOURCE',\n",
       " 'HGNC_ID',\n",
       " 'CANONICAL',\n",
       " 'MANE_SELECT',\n",
       " 'MANE_PLUS_CLINICAL',\n",
       " 'ENSP',\n",
       " 'SWISSPROT',\n",
       " 'TREMBL',\n",
       " 'UNIPARC',\n",
       " 'UNIPROT_ISOFORM',\n",
       " 'RefSeq',\n",
       " 'SIFT',\n",
       " 'PolyPhen',\n",
       " 'DOMAINS',\n",
       " 'CLIN_SIG',\n",
       " 'SOMATIC',\n",
       " 'PHENO',\n",
       " 'Conservation',\n",
       " 'BLOSUM62',\n",
       " 'GENEINFO_NAME',\n",
       " 'GENEINFO_ID',\n",
       " 'RefSeq_noversion',\n",
       " 'UniProt_IDs',\n",
       " 'PDB_path',\n",
       " 'Residue_position',\n",
       " 'Residue',\n",
       " 'DSSP_path',\n",
       " 'intra_contacts',\n",
       " 'is_catalytic',\n",
       " 'pLDDT',\n",
       " 'secondary_structure',\n",
       " 'accessibility',\n",
       " 'total energy',\n",
       " 'Backbone Hbond',\n",
       " 'Sidechain Hbond',\n",
       " 'Van der Waals',\n",
       " 'Electrostatics',\n",
       " 'Solvation Polar',\n",
       " 'Solvation Hydrophobic',\n",
       " 'Van der Waals clashes',\n",
       " 'entropy sidechain',\n",
       " 'entropy mainchain',\n",
       " 'sloop_entropy',\n",
       " 'mloop_entropy',\n",
       " 'cis_bond',\n",
       " 'torsional clash',\n",
       " 'backbone clash',\n",
       " 'helix dipole',\n",
       " 'water bridge',\n",
       " 'disulfide',\n",
       " 'electrostatic kon',\n",
       " 'partial covalent bonds',\n",
       " 'energy Ionisation',\n",
       " 'Entropy Complex']"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinvar_variants.columns.values.tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ef377f8e-21a3-4ca7-bd6a-581d026334e7",
   "metadata": {},
   "outputs": [],
   "source": [
    "clinvar_variants['intra_contacts'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "0addcef3-2857-46ff-afd1-03594ddccf25",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2     3367\n",
       "X     2953\n",
       "1     2582\n",
       "17    2291\n",
       "19    1855\n",
       "3     1774\n",
       "12    1772\n",
       "15    1564\n",
       "11    1551\n",
       "7     1537\n",
       "16    1327\n",
       "5     1198\n",
       "9     1167\n",
       "6     1056\n",
       "14     957\n",
       "10     929\n",
       "4      743\n",
       "20     638\n",
       "8      631\n",
       "13     626\n",
       "18     453\n",
       "21     264\n",
       "Y       17\n",
       "Name: #CHROM, dtype: int64"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinvar_variants['#CHROM'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "7a233bcb-2911-4b15-a1f7-1272ce5bc3c3",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0      4972\n",
       "1      1489\n",
       "2       929\n",
       "3       717\n",
       "4       699\n",
       "5       609\n",
       "7       479\n",
       "6       474\n",
       "9       431\n",
       "8       415\n",
       "13      400\n",
       "12      385\n",
       "10      375\n",
       "11      370\n",
       "15      330\n",
       "14      313\n",
       "17      297\n",
       "16      297\n",
       "18      288\n",
       "20      281\n",
       "19      269\n",
       "22      268\n",
       "21      261\n",
       "30      255\n",
       "23      244\n",
       "27      239\n",
       "31      231\n",
       "29      231\n",
       "24      229\n",
       "28      223\n",
       "25      216\n",
       "35      215\n",
       "32      206\n",
       "34      200\n",
       "39      196\n",
       "69      195\n",
       "33      193\n",
       "38      192\n",
       "26      191\n",
       "40      191\n",
       "45      188\n",
       "47      188\n",
       "41      178\n",
       "66      177\n",
       "60      174\n",
       "37      174\n",
       "42      174\n",
       "73      173\n",
       "36      173\n",
       "61      172\n",
       "62      170\n",
       "65      170\n",
       "76      169\n",
       "57      168\n",
       "71      166\n",
       "52      164\n",
       "64      164\n",
       "43      163\n",
       "63      163\n",
       "54      163\n",
       "72      162\n",
       "44      162\n",
       "68      161\n",
       "46      160\n",
       "70      159\n",
       "80      156\n",
       "59      153\n",
       "74      152\n",
       "56      149\n",
       "51      149\n",
       "49      149\n",
       "67      148\n",
       "55      145\n",
       "53      141\n",
       "50      139\n",
       "58      138\n",
       "75      136\n",
       "48      136\n",
       "77      136\n",
       "81      129\n",
       "78      105\n",
       "82      104\n",
       "84      101\n",
       "79      101\n",
       "96       98\n",
       "93       96\n",
       "87       95\n",
       "90       94\n",
       "91       91\n",
       "83       89\n",
       "103      88\n",
       "85       88\n",
       "108      87\n",
       "94       86\n",
       "113      86\n",
       "97       85\n",
       "95       85\n",
       "86       85\n",
       "98       84\n",
       "99       83\n",
       "88       83\n",
       "107      80\n",
       "104      77\n",
       "110      76\n",
       "89       75\n",
       "92       72\n",
       "100      72\n",
       "121      70\n",
       "106      70\n",
       "109      69\n",
       "119      68\n",
       "111      68\n",
       "123      68\n",
       "117      66\n",
       "105      66\n",
       "102      64\n",
       "101      61\n",
       "137      60\n",
       "124      59\n",
       "126      56\n",
       "114      55\n",
       "138      54\n",
       "115      54\n",
       "120      53\n",
       "116      52\n",
       "133      50\n",
       "118      49\n",
       "132      49\n",
       "128      48\n",
       "122      48\n",
       "131      47\n",
       "112      47\n",
       "125      45\n",
       "144      44\n",
       "130      43\n",
       "149      43\n",
       "129      39\n",
       "127      38\n",
       "134      38\n",
       "152      38\n",
       "135      38\n",
       "166      37\n",
       "143      37\n",
       "154      36\n",
       "162      34\n",
       "148      34\n",
       "140      34\n",
       "157      32\n",
       "139      32\n",
       "150      31\n",
       "136      31\n",
       "146      30\n",
       "151      30\n",
       "153      30\n",
       "156      30\n",
       "141      30\n",
       "167      29\n",
       "142      29\n",
       "174      28\n",
       "163      27\n",
       "155      26\n",
       "170      25\n",
       "158      25\n",
       "145      24\n",
       "147      24\n",
       "161      23\n",
       "165      23\n",
       "169      22\n",
       "160      22\n",
       "171      22\n",
       "175      22\n",
       "173      22\n",
       "177      21\n",
       "176      21\n",
       "179      20\n",
       "189      20\n",
       "164      20\n",
       "185      19\n",
       "184      19\n",
       "182      19\n",
       "168      18\n",
       "159      18\n",
       "200      18\n",
       "183      18\n",
       "210      17\n",
       "181      17\n",
       "180      17\n",
       "198      17\n",
       "186      16\n",
       "213      15\n",
       "191      15\n",
       "199      15\n",
       "222      15\n",
       "178      15\n",
       "190      14\n",
       "172      14\n",
       "196      14\n",
       "193      14\n",
       "192      13\n",
       "188      13\n",
       "211      12\n",
       "224      12\n",
       "187      12\n",
       "197      11\n",
       "221       9\n",
       "245       9\n",
       "201       9\n",
       "194       9\n",
       "208       8\n",
       "215       8\n",
       "204       8\n",
       "238       8\n",
       "231       8\n",
       "244       8\n",
       "237       7\n",
       "220       7\n",
       "227       7\n",
       "203       7\n",
       "242       7\n",
       "223       7\n",
       "218       7\n",
       "240       6\n",
       "207       6\n",
       "195       6\n",
       "219       6\n",
       "239       6\n",
       "233       6\n",
       "226       6\n",
       "229       6\n",
       "216       6\n",
       "241       5\n",
       "202       5\n",
       "228       5\n",
       "214       5\n",
       "206       5\n",
       "232       4\n",
       "235       4\n",
       "254       4\n",
       "243       4\n",
       "248       4\n",
       "234       4\n",
       "217       4\n",
       "236       4\n",
       "261       3\n",
       "205       3\n",
       "225       3\n",
       "252       3\n",
       "307       3\n",
       "253       3\n",
       "265       3\n",
       "212       3\n",
       "246       2\n",
       "249       2\n",
       "258       2\n",
       "260       2\n",
       "250       1\n",
       "247       1\n",
       "209       1\n",
       "251       1\n",
       "259       1\n",
       "257       1\n",
       "230       1\n",
       "Name: accessibility, dtype: int64"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinvar_variants['accessibility'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "6923f2b3-79ff-4c16-b4d7-a2c1756e4bde",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "#CHROM                                                                    Y\n",
       "POS                                                                 2787392\n",
       "ID                                                                  1470948\n",
       "REF                                                                       G\n",
       "ALT                                                                       T\n",
       "AF_ESP                                                                  NaN\n",
       "AF_EXAC                                                                 NaN\n",
       "AF_TGP                                                                  NaN\n",
       "ALLELEID                                                            1395175\n",
       "CLNDN                                                  46,XY_sex_reversal_1\n",
       "CLNDNINCL                                                               NaN\n",
       "CLNDISDB                  MONDO:MONDO:0020712,MedGen:C2748896,OMIM:40004...\n",
       "CLNDISDBINCL                                                            NaN\n",
       "CLNHGVS                                           NC_000024.10:g.2787392G>T\n",
       "CLNREVSTAT                              criteria_provided,_single_submitter\n",
       "CLNSIG                                                    Likely_pathogenic\n",
       "CLNSIGCONF                                                              NaN\n",
       "CLNSIGINCL                                                              NaN\n",
       "CLNVC                                             single_nucleotide_variant\n",
       "CLNVCSO                                                          SO:0001483\n",
       "CLNVI                                                                   NaN\n",
       "DBVARID                                                                 NaN\n",
       "MC                                              SO:0001583|missense_variant\n",
       "ORIGIN                                                                    1\n",
       "RS                                                                      NaN\n",
       "Allele                                                                    T\n",
       "Consequence                                                missense_variant\n",
       "IMPACT                                                             MODERATE\n",
       "SYMBOL                                                                  SRY\n",
       "Gene                                                        ENSG00000184895\n",
       "Feature_type                                                     Transcript\n",
       "Feature                                                     ENST00000383070\n",
       "BIOTYPE                                                      protein_coding\n",
       "EXON                                                                    1/1\n",
       "INTRON                                                                  NaN\n",
       "HGVSc                                                                   NaN\n",
       "HGVSp                                                                   NaN\n",
       "cDNA_position                                                       291/828\n",
       "CDS_position                                                        212/615\n",
       "Protein_position                                                     71/204\n",
       "Amino_acids                                                             S/Y\n",
       "Codons                                                              tCt/tAt\n",
       "Existing_variation                                                      NaN\n",
       "DISTANCE                                                                NaN\n",
       "STRAND                                                                   -1\n",
       "VARIANT_CLASS                                                           SNV\n",
       "SYMBOL_SOURCE                                                          HGNC\n",
       "HGNC_ID                                                          HGNC:11311\n",
       "CANONICAL                                                               YES\n",
       "MANE_SELECT                                                     NM_003140.3\n",
       "MANE_PLUS_CLINICAL                                                      NaN\n",
       "ENSP                                                        ENSP00000372547\n",
       "SWISSPROT                                                        Q05066.216\n",
       "TREMBL                                                           A7WPU8.138\n",
       "UNIPARC                                                       UPI0000135F78\n",
       "UNIPROT_ISOFORM                                                         NaN\n",
       "RefSeq                                                          NM_003140.3\n",
       "SIFT                                                         deleterious(0)\n",
       "PolyPhen                                                                NaN\n",
       "DOMAINS                   Gene3D:1.10.30.10&PDB-ENSP_mappings:1hry.A&PDB...\n",
       "CLIN_SIG                                                                NaN\n",
       "SOMATIC                                                                 NaN\n",
       "PHENO                                                                   NaN\n",
       "Conservation                                                            NaN\n",
       "BLOSUM62                                                                 -2\n",
       "GENEINFO_NAME                                                           SRY\n",
       "GENEINFO_ID                                                            6736\n",
       "RefSeq_noversion                                                  NM_003140\n",
       "UniProt_IDs                                                [Q05066, A7WPU8]\n",
       "PDB_path                  /ibex/scratch/projects/c2102/databases/alphafo...\n",
       "Residue_position                                                         71\n",
       "Residue                                                                 Ser\n",
       "DSSP_path                 /ibex/scratch/projects/c2102/databases/dssp/al...\n",
       "intra_contacts                                                         [94]\n",
       "is_catalytic                                                          False\n",
       "pLDDT                                                             97.966003\n",
       "secondary_structure                                                       H\n",
       "accessibility                                                             0\n",
       "total energy                                                        5.14551\n",
       "Backbone Hbond                                                    -0.283392\n",
       "Sidechain Hbond                                                     2.05072\n",
       "Van der Waals                                                      0.120522\n",
       "Electrostatics                                                    -0.062247\n",
       "Solvation Polar                                                    -1.85029\n",
       "Solvation Hydrophobic                                              -1.50452\n",
       "Van der Waals clashes                                               6.39873\n",
       "entropy sidechain                                                  -1.40358\n",
       "entropy mainchain                                                  0.548782\n",
       "sloop_entropy                                                           0.0\n",
       "mloop_entropy                                                           0.0\n",
       "cis_bond                                                                0.0\n",
       "torsional clash                                                     1.31395\n",
       "backbone clash                                                     0.342043\n",
       "helix dipole                                                      -0.183152\n",
       "water bridge                                                            0.0\n",
       "disulfide                                                               0.0\n",
       "electrostatic kon                                                       0.0\n",
       "partial covalent bonds                                                  0.0\n",
       "energy Ionisation                                                       0.0\n",
       "Entropy Complex                                                         0.0\n",
       "Name: 31245, dtype: object"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinvar_variants.loc[31245]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "218b512b-9df0-4024-9975-bd35219d9047",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['POS', 'ID', 'HGVSc', 'HGVSp', 'Conservation', 'BLOSUM62',\n",
      "       'Residue_position', 'pLDDT', 'accessibility', 'total energy',\n",
      "       'Backbone Hbond', 'Sidechain Hbond', 'Van der Waals', 'Electrostatics',\n",
      "       'Solvation Polar', 'Solvation Hydrophobic', 'Van der Waals clashes',\n",
      "       'entropy sidechain', 'entropy mainchain', 'sloop_entropy',\n",
      "       'mloop_entropy', 'cis_bond', 'torsional clash', 'backbone clash',\n",
      "       'helix dipole', 'water bridge', 'disulfide', 'electrostatic kon',\n",
      "       'partial covalent bonds', 'energy Ionisation', 'Entropy Complex'],\n",
      "      dtype='object')\n"
     ]
    }
   ],
   "source": [
    "numeric_columns = clinvar_variants.select_dtypes(include=['int', 'float']).columns\n",
    "print(numeric_columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "805d28d0-c8f2-4648-abd8-4043e8dc1011",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "7481"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinvar_variants['secondary_structure'].isnull().sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "0206c4ff-0752-4211-b332-08c33e4d22b4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "single_nucleotide_variant    31058\n",
       "Indel                          176\n",
       "Inversion                       18\n",
       "Name: CLNVC, dtype: int64"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinvar_variants['CLNVC'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "26548905-cf89-4561-a6ad-e1afdee73a47",
   "metadata": {},
   "outputs": [],
   "source": [
    "clinvar_filtered = clinvar_variants[clinvar_variants['CLNVC']=='single_nucleotide_variant'].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "1d1f60a3-3ef9-45bf-834a-f73364fb5bde",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "single_nucleotide_variant    31058\n",
       "Name: CLNVC, dtype: int64"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinvar_filtered['CLNVC'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "4f90ae77-f6f5-4a63-a320-715e2e2efe73",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1287"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinvar_filtered['Conservation'].isnull().sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "a97edeb6-0841-42b9-a4f9-b63aeae1e8cf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "30732"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinvar_filtered['AF_TGP'].isnull().sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "8572b5ce-fd54-4791-a5cd-4b93b7fc4fa9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.0002     238\n",
       "0.0004      36\n",
       "0.0006      13\n",
       "0.0008       6\n",
       "0.001        5\n",
       "0.00053      3\n",
       "0.0024       3\n",
       "0            3\n",
       "0.0014       2\n",
       "0.002        2\n",
       "0.00026      2\n",
       "0.0012       2\n",
       "0.0018       1\n",
       "0.00319      1\n",
       "0.01358      1\n",
       "0.00359      1\n",
       "0.0016       1\n",
       "0.28894      1\n",
       "0.01478      1\n",
       "0.45647      1\n",
       "0.00459      1\n",
       "0.03603      1\n",
       "0.02396      1\n",
       "Name: AF_TGP, dtype: int64"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinvar_filtered['AF_TGP'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "f9e4931f-b0ac-48c4-9079-07d1867f76fe",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Early_infantile_epileptic_encephalopathy_with_suppression_bursts                                                                                                                                                                                                           422\n",
       "Hypercholesterolemia,_familial,_1                                                                                                                                                                                                                                          367\n",
       "Inborn_genetic_diseases                                                                                                                                                                                                                                                    354\n",
       "Ehlers-Danlos_syndrome,_type_4                                                                                                                                                                                                                                             310\n",
       "Marfan_syndrome|Familial_thoracic_aortic_aneurysm_and_aortic_dissection                                                                                                                                                                                                    246\n",
       "Marfan_syndrome                                                                                                                                                                                                                                                            235\n",
       "Familial_thoracic_aortic_aneurysm_and_aortic_dissection|Marfan_syndrome                                                                                                                                                                                                    212\n",
       "                                                                                                                                                                                                                                                                          ... \n",
       "Spermatogenic_failure_64                                                                                                                                                                                                                                                     1\n",
       "Charcot-Marie-Tooth_disease|Charcot-Marie-Tooth_disease_axonal_type_2K|Charcot-Marie-Tooth_disease_type_4A                                                                                                                                                                   1\n",
       "Charcot-Marie-Tooth_disease_axonal_type_2K|Charcot-Marie-Tooth_disease_type_4A                                                                                                                                                                                               1\n",
       "Charcot-Marie-Tooth_disease|Charcot-Marie-Tooth_disease_axonal_type_2K|Charcot-Marie-Tooth_disease_type_4A|Charcot-Marie-Tooth_disease,_axonal,_with_vocal_cord_paresis,_autosomal_recessive|Charcot-Marie-Tooth_disease_recessive_intermediate_A|Peripheral_neuropathy      1\n",
       "Charcot-Marie-Tooth_disease|Charcot-Marie-Tooth_disease_type_4A                                                                                                                                                                                                              1\n",
       "Familial_episodic_pain_syndrome_with_predominantly_upper_body_involvement                                                                                                                                                                                                    1\n",
       "Deafness,_Y-linked_2                                                                                                                                                                                                                                                         1\n",
       "Name: CLNDN, Length: 7917, dtype: int64"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinvar_filtered['CLNDN'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "f0b59904-4492-4ed0-ac66-cc2d70195882",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(31058, 100)"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinvar_filtered.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "33dbebb5-a9e5-4d2b-b065-297ae9dd57be",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define the keywords to search for\n",
    "keywords = ['heart', 'cardiovascular', 'cancer', 'carcinoma', 'sarcoma', 'leuchemia', 'lymphoma', 'myeloma']\n",
    "\n",
    "# Add a new column based on the annotation\n",
    "clinvar_filtered['DiseaseClass'] = clinvar_filtered['CLNDN'].apply(lambda x: 'heart disease' if any(keyword in str(x).lower() for keyword in keywords[:2]) else ('cancer' if 'cancer' in str(x).lower() else ('carcinoma' if 'carcinoma' in str(x).lower() else ('sarcoma' if 'sarcoma' in str(x).lower() else ('leuchemia' if 'leuchemia' in str(x).lower() else ('lymphoma' if 'lymphoma' in str(x).lower() else ('myeloma' if 'myeloma' in str(x).lower() else None)))))))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "34fef192-5715-4001-95e8-6212db1dff23",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "cancer           820\n",
       "heart disease    420\n",
       "carcinoma        414\n",
       "myeloma           34\n",
       "lymphoma          24\n",
       "sarcoma            7\n",
       "Name: DiseaseClass, dtype: int64"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinvar_filtered['DiseaseClass'].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8e1a86c9-fcd7-4b19-a315-0e54e4d290bd",
   "metadata": {},
   "source": [
    "### add polarities & change in size & add number of contacts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "a2bff0fb-a2e4-4466-ac43-b1f402d3a9e8",
   "metadata": {},
   "outputs": [],
   "source": [
    "clinvar_filtered[['WT', 'MUT']] = clinvar_filtered['Amino_acids'].str.split('/', expand=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "82f4688b-8746-4998-9bc2-f9ebc62caae1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Convert one-letter code to three-letter code\n",
    "clinvar_filtered['WT'] = clinvar_filtered['WT'].map(protein_letters_1to3)\n",
    "clinvar_filtered['MUT'] = clinvar_filtered['MUT'].map(protein_letters_1to3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "844e743c-a206-4c84-bd62-b5bff6609ce0",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "#CHROM                                                                    Y\n",
       "POS                                                                 2787392\n",
       "ID                                                                  1470948\n",
       "REF                                                                       G\n",
       "ALT                                                                       T\n",
       "AF_ESP                                                                  NaN\n",
       "AF_EXAC                                                                 NaN\n",
       "AF_TGP                                                                  NaN\n",
       "ALLELEID                                                            1395175\n",
       "CLNDN                                                  46,XY_sex_reversal_1\n",
       "CLNDNINCL                                                               NaN\n",
       "CLNDISDB                  MONDO:MONDO:0020712,MedGen:C2748896,OMIM:40004...\n",
       "CLNDISDBINCL                                                            NaN\n",
       "CLNHGVS                                           NC_000024.10:g.2787392G>T\n",
       "CLNREVSTAT                              criteria_provided,_single_submitter\n",
       "CLNSIG                                                    Likely_pathogenic\n",
       "CLNSIGCONF                                                              NaN\n",
       "CLNSIGINCL                                                              NaN\n",
       "CLNVC                                             single_nucleotide_variant\n",
       "CLNVCSO                                                          SO:0001483\n",
       "CLNVI                                                                   NaN\n",
       "DBVARID                                                                 NaN\n",
       "MC                                              SO:0001583|missense_variant\n",
       "ORIGIN                                                                    1\n",
       "RS                                                                      NaN\n",
       "Allele                                                                    T\n",
       "Consequence                                                missense_variant\n",
       "IMPACT                                                             MODERATE\n",
       "SYMBOL                                                                  SRY\n",
       "Gene                                                        ENSG00000184895\n",
       "Feature_type                                                     Transcript\n",
       "Feature                                                     ENST00000383070\n",
       "BIOTYPE                                                      protein_coding\n",
       "EXON                                                                    1/1\n",
       "INTRON                                                                  NaN\n",
       "HGVSc                                                                   NaN\n",
       "HGVSp                                                                   NaN\n",
       "cDNA_position                                                       291/828\n",
       "CDS_position                                                        212/615\n",
       "Protein_position                                                     71/204\n",
       "Amino_acids                                                             S/Y\n",
       "Codons                                                              tCt/tAt\n",
       "Existing_variation                                                      NaN\n",
       "DISTANCE                                                                NaN\n",
       "STRAND                                                                   -1\n",
       "VARIANT_CLASS                                                           SNV\n",
       "SYMBOL_SOURCE                                                          HGNC\n",
       "HGNC_ID                                                          HGNC:11311\n",
       "CANONICAL                                                               YES\n",
       "MANE_SELECT                                                     NM_003140.3\n",
       "MANE_PLUS_CLINICAL                                                      NaN\n",
       "ENSP                                                        ENSP00000372547\n",
       "SWISSPROT                                                        Q05066.216\n",
       "TREMBL                                                           A7WPU8.138\n",
       "UNIPARC                                                       UPI0000135F78\n",
       "UNIPROT_ISOFORM                                                         NaN\n",
       "RefSeq                                                          NM_003140.3\n",
       "SIFT                                                         deleterious(0)\n",
       "PolyPhen                                                                NaN\n",
       "DOMAINS                   Gene3D:1.10.30.10&PDB-ENSP_mappings:1hry.A&PDB...\n",
       "CLIN_SIG                                                                NaN\n",
       "SOMATIC                                                                 NaN\n",
       "PHENO                                                                   NaN\n",
       "Conservation                                                            NaN\n",
       "BLOSUM62                                                                 -2\n",
       "GENEINFO_NAME                                                           SRY\n",
       "GENEINFO_ID                                                            6736\n",
       "RefSeq_noversion                                                  NM_003140\n",
       "UniProt_IDs                                                [Q05066, A7WPU8]\n",
       "PDB_path                  /ibex/scratch/projects/c2102/databases/alphafo...\n",
       "Residue_position                                                         71\n",
       "Residue                                                                 Ser\n",
       "DSSP_path                 /ibex/scratch/projects/c2102/databases/dssp/al...\n",
       "intra_contacts                                                         [94]\n",
       "is_catalytic                                                          False\n",
       "pLDDT                                                             97.966003\n",
       "secondary_structure                                                       H\n",
       "accessibility                                                             0\n",
       "total energy                                                        5.14551\n",
       "Backbone Hbond                                                    -0.283392\n",
       "Sidechain Hbond                                                     2.05072\n",
       "Van der Waals                                                      0.120522\n",
       "Electrostatics                                                    -0.062247\n",
       "Solvation Polar                                                    -1.85029\n",
       "Solvation Hydrophobic                                              -1.50452\n",
       "Van der Waals clashes                                               6.39873\n",
       "entropy sidechain                                                  -1.40358\n",
       "entropy mainchain                                                  0.548782\n",
       "sloop_entropy                                                           0.0\n",
       "mloop_entropy                                                           0.0\n",
       "cis_bond                                                                0.0\n",
       "torsional clash                                                     1.31395\n",
       "backbone clash                                                     0.342043\n",
       "helix dipole                                                      -0.183152\n",
       "water bridge                                                            0.0\n",
       "disulfide                                                               0.0\n",
       "electrostatic kon                                                       0.0\n",
       "partial covalent bonds                                                  0.0\n",
       "energy Ionisation                                                       0.0\n",
       "Entropy Complex                                                         0.0\n",
       "DiseaseClass                                                           None\n",
       "WT                                                                      Ser\n",
       "MUT                                                                     Tyr\n",
       "Name: 31245, dtype: object"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinvar_filtered.loc[31245]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "c1629038-3b56-439b-86a0-93270397986b",
   "metadata": {},
   "outputs": [],
   "source": [
    "aliphatic = [\n",
    "    'Ala',\n",
    "    'Ile',\n",
    "    'Leu',\n",
    "    'Met',\n",
    "    'Val'\n",
    "]\n",
    "\n",
    "aromatic = [\n",
    "    'Phe',\n",
    "    'Trp',\n",
    "    'Tyr'\n",
    "]\n",
    "\n",
    "hydrophobic = aliphatic + aromatic\n",
    "\n",
    "positive = [\n",
    "    'Arg',\n",
    "    'His',\n",
    "    'Lys'\n",
    "]\n",
    "\n",
    "negative = [\n",
    "    'Asp',\n",
    "    'Glu'\n",
    "]\n",
    "\n",
    "neutral = [\n",
    "    'Asn',\n",
    "    'Gln',\n",
    "    'Thr',\n",
    "    'Ser'\n",
    "]\n",
    "\n",
    "polar = positive + negative + neutral\n",
    "\n",
    "special = [\n",
    "    'Pro',\n",
    "    'Gly',\n",
    "    'Cys'\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "eae9bf01-b447-4b4d-ad78-d3cc601450a3",
   "metadata": {},
   "outputs": [],
   "source": [
    "clinvar_filtered['aliphatic_assign_WT'] = clinvar_filtered.WT.isin(aliphatic).apply(\n",
    "                            lambda x: 1 if x else 0)\n",
    "features_vector = []\n",
    "clinvar_filtered['aromatic_assign_WT'] = clinvar_filtered.WT.isin(aromatic).apply(\n",
    "                            lambda x: 1 if x else 0)\n",
    "clinvar_filtered['hydrophobic_assign_WT'] = clinvar_filtered.WT.isin(hydrophobic).apply(\n",
    "                            lambda x: 1 if x else 0)\n",
    "clinvar_filtered['positive_assign_WT'] = clinvar_filtered.WT.isin(positive).apply(\n",
    "                            lambda x: 1 if x else 0)\n",
    "clinvar_filtered['negative_assign_WT'] = clinvar_filtered.WT.isin(negative).apply(\n",
    "                            lambda x: 1 if x else 0)\n",
    "clinvar_filtered['neutral_assign_WT'] = clinvar_filtered.WT.isin(neutral).apply(\n",
    "                            lambda x: 1 if x else 0)\n",
    "clinvar_filtered['polar_assign_WT'] = clinvar_filtered.WT.isin(polar).apply(\n",
    "                            lambda x: 1 if x else 0)\n",
    "clinvar_filtered['special_assign_WT'] = clinvar_filtered.WT.isin(special).apply(\n",
    "                            lambda x: 1 if x else 0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "88018307-6dc0-48db-89bc-6f26bce96c7a",
   "metadata": {},
   "outputs": [],
   "source": [
    "clinvar_filtered['aliphatic_assign_MUT'] = clinvar_filtered.MUT.isin(aliphatic).apply(\n",
    "                            lambda x: 1 if x else 0)\n",
    "features_vector = []\n",
    "clinvar_filtered['aromatic_assign_MUT'] = clinvar_filtered.MUT.isin(aromatic).apply(\n",
    "                            lambda x: 1 if x else 0)\n",
    "clinvar_filtered['hydrophobic_assign_MUT'] = clinvar_filtered.MUT.isin(hydrophobic).apply(\n",
    "                            lambda x: 1 if x else 0)\n",
    "clinvar_filtered['positive_assign_MUT'] = clinvar_filtered.MUT.isin(positive).apply(\n",
    "                            lambda x: 1 if x else 0)\n",
    "clinvar_filtered['negative_assign_MUT'] = clinvar_filtered.MUT.isin(negative).apply(\n",
    "                            lambda x: 1 if x else 0)\n",
    "clinvar_filtered['neutral_assign_MUT'] = clinvar_filtered.MUT.isin(neutral).apply(\n",
    "                            lambda x: 1 if x else 0)\n",
    "clinvar_filtered['polar_assign_MUT'] = clinvar_filtered.MUT.isin(polar).apply(\n",
    "                            lambda x: 1 if x else 0)\n",
    "clinvar_filtered['special_assign_MUT'] = clinvar_filtered.MUT.isin(special).apply(\n",
    "                            lambda x: 1 if x else 0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "id": "18052d76-d383-438a-a029-303b53a1381f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><div id=a5e717c2-1470-41d5-8656-6e6480f744ad style=\"display:none; background-color:#9D6CFF; color:white; width:200px; height:30px; padding-left:5px; border-radius:4px; flex-direction:row; justify-content:space-around; align-items:center;\" onmouseover=\"this.style.backgroundColor='#BA9BF8'\" onmouseout=\"this.style.backgroundColor='#9D6CFF'\" onclick=\"window.commands?.execute('create-mitosheet-from-dataframe-output');\">See Full Dataframe in Mito</div> <script> if (window.commands?.hasCommand('create-mitosheet-from-dataframe-output')) document.getElementById('a5e717c2-1470-41d5-8656-6e6480f744ad').style.display = 'flex' </script> <table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>AF_ESP</th>\n",
       "      <th>AF_EXAC</th>\n",
       "      <th>AF_TGP</th>\n",
       "      <th>ALLELEID</th>\n",
       "      <th>CLNDN</th>\n",
       "      <th>...</th>\n",
       "      <th>polar_assign_WT</th>\n",
       "      <th>special_assign_WT</th>\n",
       "      <th>aliphatic_assign_MUT</th>\n",
       "      <th>aromatic_assign_MUT</th>\n",
       "      <th>hydrophobic_assign_MUT</th>\n",
       "      <th>positive_assign_MUT</th>\n",
       "      <th>negative_assign_MUT</th>\n",
       "      <th>neutral_assign_MUT</th>\n",
       "      <th>polar_assign_MUT</th>\n",
       "      <th>special_assign_MUT</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>197429459</td>\n",
       "      <td>1048144</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1036051</td>\n",
       "      <td>Leber_congenital_amaurosis_8</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>197429466</td>\n",
       "      <td>1213987</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1203980</td>\n",
       "      <td>Retinitis_pigmentosa_12</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>197429467</td>\n",
       "      <td>1963628</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2017299</td>\n",
       "      <td>Retinitis_pigmentosa_12|Leber_congenital_amaurosis_8</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>197429468</td>\n",
       "      <td>2202910</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1928560</td>\n",
       "      <td>Retinitis_pigmentosa_12|Leber_congenital_amaurosis_8</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>197429473</td>\n",
       "      <td>978993</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>967053</td>\n",
       "      <td>Autosomal_recessive_retinitis_pigmentosa</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1</td>\n",
       "      <td>197429555</td>\n",
       "      <td>191357</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>189156</td>\n",
       "      <td>Retinitis_pigmentosa</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1</td>\n",
       "      <td>197429581</td>\n",
       "      <td>813170</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>801275</td>\n",
       "      <td>Leber_congenital_amaurosis</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31051</th>\n",
       "      <td>Y</td>\n",
       "      <td>2787392</td>\n",
       "      <td>1470948</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1395175</td>\n",
       "      <td>46,XY_sex_reversal_1</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31052</th>\n",
       "      <td>Y</td>\n",
       "      <td>2787401</td>\n",
       "      <td>9746</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>24785</td>\n",
       "      <td>46,XY_sex_reversal_1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31053</th>\n",
       "      <td>Y</td>\n",
       "      <td>2787408</td>\n",
       "      <td>1177296</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1166601</td>\n",
       "      <td>46,XY_disorder_of_sex_development</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31054</th>\n",
       "      <td>Y</td>\n",
       "      <td>2787412</td>\n",
       "      <td>9747</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>24786</td>\n",
       "      <td>46,XY_sex_reversal_1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31055</th>\n",
       "      <td>Y</td>\n",
       "      <td>2787515</td>\n",
       "      <td>492908</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>485861</td>\n",
       "      <td>46,XY_sex_reversal_1</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31056</th>\n",
       "      <td>Y</td>\n",
       "      <td>2787551</td>\n",
       "      <td>9754</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>NaN</td>\n",
       "      <td>9e-05</td>\n",
       "      <td>NaN</td>\n",
       "      <td>24793</td>\n",
       "      <td>46,XY_sex_reversal_1</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31057</th>\n",
       "      <td>Y</td>\n",
       "      <td>7063898</td>\n",
       "      <td>625467</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>0.00123</td>\n",
       "      <td>0.00048</td>\n",
       "      <td>NaN</td>\n",
       "      <td>613772</td>\n",
       "      <td>Deafness,_Y-linked_2</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table></div>"
      ],
      "text/plain": [
       "      #CHROM        POS       ID REF ALT   AF_ESP  AF_EXAC AF_TGP ALLELEID  \\\n",
       "0          1  197429459  1048144   G   C      NaN      NaN    NaN  1036051   \n",
       "1          1  197429466  1213987   T   G      NaN      NaN    NaN  1203980   \n",
       "2          1  197429467  1963628   G   A      NaN      NaN    NaN  2017299   \n",
       "3          1  197429468  2202910   G   C      NaN      NaN    NaN  1928560   \n",
       "4          1  197429473   978993   G   T      NaN      NaN    NaN   967053   \n",
       "5          1  197429555   191357   G   A      NaN      NaN    NaN   189156   \n",
       "6          1  197429581   813170   G   C      NaN      NaN    NaN   801275   \n",
       "...      ...        ...      ...  ..  ..      ...      ...    ...      ...   \n",
       "31051      Y    2787392  1470948   G   T      NaN      NaN    NaN  1395175   \n",
       "31052      Y    2787401     9746   A   G      NaN      NaN    NaN    24785   \n",
       "31053      Y    2787408  1177296   C   G      NaN      NaN    NaN  1166601   \n",
       "31054      Y    2787412     9747   C   T      NaN      NaN    NaN    24786   \n",
       "31055      Y    2787515   492908   C   A      NaN      NaN    NaN   485861   \n",
       "31056      Y    2787551     9754   C   T      NaN    9e-05    NaN    24793   \n",
       "31057      Y    7063898   625467   A   T  0.00123  0.00048    NaN   613772   \n",
       "\n",
       "                                                   CLNDN  ... polar_assign_WT  \\\n",
       "0                           Leber_congenital_amaurosis_8  ...               0   \n",
       "1                                Retinitis_pigmentosa_12  ...               1   \n",
       "2      Retinitis_pigmentosa_12|Leber_congenital_amaur...  ...               0   \n",
       "3      Retinitis_pigmentosa_12|Leber_congenital_amaur...  ...               0   \n",
       "4               Autosomal_recessive_retinitis_pigmentosa  ...               0   \n",
       "5                                   Retinitis_pigmentosa  ...               0   \n",
       "6                             Leber_congenital_amaurosis  ...               0   \n",
       "...                                                  ...  ...             ...   \n",
       "31051                               46,XY_sex_reversal_1  ...               1   \n",
       "31052                               46,XY_sex_reversal_1  ...               0   \n",
       "31053                  46,XY_disorder_of_sex_development  ...               0   \n",
       "31054                               46,XY_sex_reversal_1  ...               0   \n",
       "31055                               46,XY_sex_reversal_1  ...               1   \n",
       "31056                               46,XY_sex_reversal_1  ...               1   \n",
       "31057                               Deafness,_Y-linked_2  ...               1   \n",
       "\n",
       "      special_assign_WT aliphatic_assign_MUT aromatic_assign_MUT  \\\n",
       "0                     1                    0                   0   \n",
       "1                     0                    0                   0   \n",
       "2                     1                    0                   0   \n",
       "3                     1                    1                   0   \n",
       "4                     0                    0                   1   \n",
       "5                     1                    0                   1   \n",
       "6                     0                    0                   0   \n",
       "...                 ...                  ...                 ...   \n",
       "31051                 0                    0                   1   \n",
       "31052                 0                    0                   0   \n",
       "31053                 0                    0                   0   \n",
       "31054                 0                    1                   0   \n",
       "31055                 0                    1                   0   \n",
       "31056                 0                    0                   0   \n",
       "31057                 0                    1                   0   \n",
       "\n",
       "      hydrophobic_assign_MUT positive_assign_MUT negative_assign_MUT  \\\n",
       "0                          0                   0                   0   \n",
       "1                          0                   1                   0   \n",
       "2                          0                   1                   0   \n",
       "3                          1                   0                   0   \n",
       "4                          1                   0                   0   \n",
       "5                          1                   0                   0   \n",
       "6                          0                   0                   0   \n",
       "...                      ...                 ...                 ...   \n",
       "31051                      1                   0                   0   \n",
       "31052                      0                   0                   0   \n",
       "31053                      0                   0                   0   \n",
       "31054                      1                   0                   0   \n",
       "31055                      1                   0                   0   \n",
       "31056                      0                   0                   0   \n",
       "31057                      1                   0                   0   \n",
       "\n",
       "      neutral_assign_MUT polar_assign_MUT special_assign_MUT  \n",
       "0                      1                1                  0  \n",
       "1                      0                1                  0  \n",
       "2                      0                1                  0  \n",
       "3                      0                0                  0  \n",
       "4                      0                0                  0  \n",
       "5                      0                0                  0  \n",
       "6                      0                0                  1  \n",
       "...                  ...              ...                ...  \n",
       "31051                  0                0                  0  \n",
       "31052                  1                1                  0  \n",
       "31053                  0                0                  1  \n",
       "31054                  0                0                  0  \n",
       "31055                  0                0                  0  \n",
       "31056                  1                1                  0  \n",
       "31057                  0                0                  0  \n",
       "\n",
       "[31058 rows x 119 columns]"
      ]
     },
     "execution_count": 89,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinvar_filtered"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "id": "dcd87712-1c3d-4e1a-b3cc-ae190dadf3bc",
   "metadata": {},
   "outputs": [],
   "source": [
    "clinvar_filtered = clinvar_filtered.reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "id": "c9384875-b323-4334-9730-660e46e25723",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "31058"
      ]
     },
     "execution_count": 90,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(clinvar_filtered)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "da1b1701-56ff-486b-a4fb-a8447991497a",
   "metadata": {},
   "source": [
    "#### add changes in polarity"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 142,
   "id": "14d54750-2ae5-4de4-90c1-0c0359de0ff5",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/56/l75lqrhd7m97zxtpbdw9rbbc0000gp/T/ipykernel_8537/498000398.py:11: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  clinvar_filtered['changes'].loc[i]=clinvar_filtered['changes'].loc[i]+prop+'_'\n",
      "/var/folders/56/l75lqrhd7m97zxtpbdw9rbbc0000gp/T/ipykernel_8537/498000398.py:16: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  clinvar_filtered['changes'].loc[i]=clinvar_filtered['changes'].loc[i]+'to_'+prop+'_'\n",
      "/var/folders/56/l75lqrhd7m97zxtpbdw9rbbc0000gp/T/ipykernel_8537/498000398.py:19: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  clinvar_filtered['changes'].loc[i]=clinvar_filtered['changes'].loc[i]+prop+'_'\n",
      "/var/folders/56/l75lqrhd7m97zxtpbdw9rbbc0000gp/T/ipykernel_8537/498000398.py:20: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  clinvar_filtered['changes'].loc[i] = clinvar_filtered['changes'].str[:-1].loc[i]\n"
     ]
    }
   ],
   "source": [
    "# Function to extract property names from column names\n",
    "def extract_property(column_name):\n",
    "    return column_name.split('_')[0]\n",
    "\n",
    "clinvar_filtered['changes']=''\n",
    "for i in range(0,len(clinvar_filtered)):\n",
    "    for column in clinvar_filtered.columns:\n",
    "        if column.endswith('_WT'):\n",
    "            if clinvar_filtered[column].loc[i] == 1:\n",
    "                prop = extract_property(column)\n",
    "                clinvar_filtered['changes'].loc[i]=clinvar_filtered['changes'].loc[i]+prop+'_'\n",
    "        elif column.endswith('_MUT'):\n",
    "            if clinvar_filtered[column].loc[i] == 1:\n",
    "                if clinvar_filtered['changes'].str.contains('to').iloc[i] == False:\n",
    "                    prop = extract_property(column)\n",
    "                    clinvar_filtered['changes'].loc[i]=clinvar_filtered['changes'].loc[i]+'to_'+prop+'_'\n",
    "                else:\n",
    "                    prop = extract_property(column)\n",
    "                    clinvar_filtered['changes'].loc[i]=clinvar_filtered['changes'].loc[i]+prop+'_'\n",
    "    clinvar_filtered['changes'].loc[i] = clinvar_filtered['changes'].str[:-1].loc[i]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "id": "7f6badde-a92f-4366-ac7a-c11ea78d9fad",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><div id=e9b111cc-324e-45db-8740-fbb2ca1d7dc9 style=\"display:none; background-color:#9D6CFF; color:white; width:200px; height:30px; padding-left:5px; border-radius:4px; flex-direction:row; justify-content:space-around; align-items:center;\" onmouseover=\"this.style.backgroundColor='#BA9BF8'\" onmouseout=\"this.style.backgroundColor='#9D6CFF'\" onclick=\"window.commands?.execute('create-mitosheet-from-dataframe-output');\">See Full Dataframe in Mito</div> <script> if (window.commands?.hasCommand('create-mitosheet-from-dataframe-output')) document.getElementById('e9b111cc-324e-45db-8740-fbb2ca1d7dc9').style.display = 'flex' </script> <table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>AF_ESP</th>\n",
       "      <th>AF_EXAC</th>\n",
       "      <th>AF_TGP</th>\n",
       "      <th>ALLELEID</th>\n",
       "      <th>CLNDN</th>\n",
       "      <th>...</th>\n",
       "      <th>special_assign_WT</th>\n",
       "      <th>aliphatic_assign_MUT</th>\n",
       "      <th>aromatic_assign_MUT</th>\n",
       "      <th>hydrophobic_assign_MUT</th>\n",
       "      <th>positive_assign_MUT</th>\n",
       "      <th>negative_assign_MUT</th>\n",
       "      <th>neutral_assign_MUT</th>\n",
       "      <th>polar_assign_MUT</th>\n",
       "      <th>special_assign_MUT</th>\n",
       "      <th>changes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>197429459</td>\n",
       "      <td>1048144</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1036051</td>\n",
       "      <td>Leber_congenital_amaurosis_8</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>special_to_neutral_polar</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>197429466</td>\n",
       "      <td>1213987</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1203980</td>\n",
       "      <td>Retinitis_pigmentosa_12</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>neutral_polar_to_positive_polar</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>197429467</td>\n",
       "      <td>1963628</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2017299</td>\n",
       "      <td>Retinitis_pigmentosa_12|Leber_congenital_amaurosis_8</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>special_to_positive_polar</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>197429468</td>\n",
       "      <td>2202910</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1928560</td>\n",
       "      <td>Retinitis_pigmentosa_12|Leber_congenital_amaurosis_8</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>special_to_aliphatic_hydrophobic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>197429473</td>\n",
       "      <td>978993</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>967053</td>\n",
       "      <td>Autosomal_recessive_retinitis_pigmentosa</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>aliphatic_hydrophobic_to_aromatic_hydrophobic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1</td>\n",
       "      <td>197429555</td>\n",
       "      <td>191357</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>189156</td>\n",
       "      <td>Retinitis_pigmentosa</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>special_to_aromatic_hydrophobic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1</td>\n",
       "      <td>197429581</td>\n",
       "      <td>813170</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>801275</td>\n",
       "      <td>Leber_congenital_amaurosis</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>aliphatic_hydrophobic_to_special</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31051</th>\n",
       "      <td>Y</td>\n",
       "      <td>2787392</td>\n",
       "      <td>1470948</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1395175</td>\n",
       "      <td>46,XY_sex_reversal_1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>neutral_polar_to_aromatic_hydrophobic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31052</th>\n",
       "      <td>Y</td>\n",
       "      <td>2787401</td>\n",
       "      <td>9746</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>24785</td>\n",
       "      <td>46,XY_sex_reversal_1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>aliphatic_hydrophobic_to_neutral_polar</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31053</th>\n",
       "      <td>Y</td>\n",
       "      <td>2787408</td>\n",
       "      <td>1177296</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1166601</td>\n",
       "      <td>46,XY_disorder_of_sex_development</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>aliphatic_hydrophobic_to_special</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31054</th>\n",
       "      <td>Y</td>\n",
       "      <td>2787412</td>\n",
       "      <td>9747</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>24786</td>\n",
       "      <td>46,XY_sex_reversal_1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>aliphatic_hydrophobic_to_aliphatic_hydrophobic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31055</th>\n",
       "      <td>Y</td>\n",
       "      <td>2787515</td>\n",
       "      <td>492908</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>485861</td>\n",
       "      <td>46,XY_sex_reversal_1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>positive_polar_to_aliphatic_hydrophobic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31056</th>\n",
       "      <td>Y</td>\n",
       "      <td>2787551</td>\n",
       "      <td>9754</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>NaN</td>\n",
       "      <td>9e-05</td>\n",
       "      <td>NaN</td>\n",
       "      <td>24793</td>\n",
       "      <td>46,XY_sex_reversal_1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>neutral_polar_to_neutral_polar</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31057</th>\n",
       "      <td>Y</td>\n",
       "      <td>7063898</td>\n",
       "      <td>625467</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>0.00123</td>\n",
       "      <td>0.00048</td>\n",
       "      <td>NaN</td>\n",
       "      <td>613772</td>\n",
       "      <td>Deafness,_Y-linked_2</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>negative_polar_to_aliphatic_hydrophobic</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table></div>"
      ],
      "text/plain": [
       "      #CHROM        POS       ID REF ALT   AF_ESP  AF_EXAC AF_TGP ALLELEID  \\\n",
       "0          1  197429459  1048144   G   C      NaN      NaN    NaN  1036051   \n",
       "1          1  197429466  1213987   T   G      NaN      NaN    NaN  1203980   \n",
       "2          1  197429467  1963628   G   A      NaN      NaN    NaN  2017299   \n",
       "3          1  197429468  2202910   G   C      NaN      NaN    NaN  1928560   \n",
       "4          1  197429473   978993   G   T      NaN      NaN    NaN   967053   \n",
       "5          1  197429555   191357   G   A      NaN      NaN    NaN   189156   \n",
       "6          1  197429581   813170   G   C      NaN      NaN    NaN   801275   \n",
       "...      ...        ...      ...  ..  ..      ...      ...    ...      ...   \n",
       "31051      Y    2787392  1470948   G   T      NaN      NaN    NaN  1395175   \n",
       "31052      Y    2787401     9746   A   G      NaN      NaN    NaN    24785   \n",
       "31053      Y    2787408  1177296   C   G      NaN      NaN    NaN  1166601   \n",
       "31054      Y    2787412     9747   C   T      NaN      NaN    NaN    24786   \n",
       "31055      Y    2787515   492908   C   A      NaN      NaN    NaN   485861   \n",
       "31056      Y    2787551     9754   C   T      NaN    9e-05    NaN    24793   \n",
       "31057      Y    7063898   625467   A   T  0.00123  0.00048    NaN   613772   \n",
       "\n",
       "                                                   CLNDN  ...  \\\n",
       "0                           Leber_congenital_amaurosis_8  ...   \n",
       "1                                Retinitis_pigmentosa_12  ...   \n",
       "2      Retinitis_pigmentosa_12|Leber_congenital_amaur...  ...   \n",
       "3      Retinitis_pigmentosa_12|Leber_congenital_amaur...  ...   \n",
       "4               Autosomal_recessive_retinitis_pigmentosa  ...   \n",
       "5                                   Retinitis_pigmentosa  ...   \n",
       "6                             Leber_congenital_amaurosis  ...   \n",
       "...                                                  ...  ...   \n",
       "31051                               46,XY_sex_reversal_1  ...   \n",
       "31052                               46,XY_sex_reversal_1  ...   \n",
       "31053                  46,XY_disorder_of_sex_development  ...   \n",
       "31054                               46,XY_sex_reversal_1  ...   \n",
       "31055                               46,XY_sex_reversal_1  ...   \n",
       "31056                               46,XY_sex_reversal_1  ...   \n",
       "31057                               Deafness,_Y-linked_2  ...   \n",
       "\n",
       "      special_assign_WT aliphatic_assign_MUT aromatic_assign_MUT  \\\n",
       "0                     1                    0                   0   \n",
       "1                     0                    0                   0   \n",
       "2                     1                    0                   0   \n",
       "3                     1                    1                   0   \n",
       "4                     0                    0                   1   \n",
       "5                     1                    0                   1   \n",
       "6                     0                    0                   0   \n",
       "...                 ...                  ...                 ...   \n",
       "31051                 0                    0                   1   \n",
       "31052                 0                    0                   0   \n",
       "31053                 0                    0                   0   \n",
       "31054                 0                    1                   0   \n",
       "31055                 0                    1                   0   \n",
       "31056                 0                    0                   0   \n",
       "31057                 0                    1                   0   \n",
       "\n",
       "      hydrophobic_assign_MUT positive_assign_MUT negative_assign_MUT  \\\n",
       "0                          0                   0                   0   \n",
       "1                          0                   1                   0   \n",
       "2                          0                   1                   0   \n",
       "3                          1                   0                   0   \n",
       "4                          1                   0                   0   \n",
       "5                          1                   0                   0   \n",
       "6                          0                   0                   0   \n",
       "...                      ...                 ...                 ...   \n",
       "31051                      1                   0                   0   \n",
       "31052                      0                   0                   0   \n",
       "31053                      0                   0                   0   \n",
       "31054                      1                   0                   0   \n",
       "31055                      1                   0                   0   \n",
       "31056                      0                   0                   0   \n",
       "31057                      1                   0                   0   \n",
       "\n",
       "      neutral_assign_MUT polar_assign_MUT special_assign_MUT  \\\n",
       "0                      1                1                  0   \n",
       "1                      0                1                  0   \n",
       "2                      0                1                  0   \n",
       "3                      0                0                  0   \n",
       "4                      0                0                  0   \n",
       "5                      0                0                  0   \n",
       "6                      0                0                  1   \n",
       "...                  ...              ...                ...   \n",
       "31051                  0                0                  0   \n",
       "31052                  1                1                  0   \n",
       "31053                  0                0                  1   \n",
       "31054                  0                0                  0   \n",
       "31055                  0                0                  0   \n",
       "31056                  1                1                  0   \n",
       "31057                  0                0                  0   \n",
       "\n",
       "                                              changes  \n",
       "0                            special_to_neutral_polar  \n",
       "1                     neutral_polar_to_positive_polar  \n",
       "2                           special_to_positive_polar  \n",
       "3                    special_to_aliphatic_hydrophobic  \n",
       "4       aliphatic_hydrophobic_to_aromatic_hydrophobic  \n",
       "5                     special_to_aromatic_hydrophobic  \n",
       "6                    aliphatic_hydrophobic_to_special  \n",
       "...                                               ...  \n",
       "31051           neutral_polar_to_aromatic_hydrophobic  \n",
       "31052          aliphatic_hydrophobic_to_neutral_polar  \n",
       "31053                aliphatic_hydrophobic_to_special  \n",
       "31054  aliphatic_hydrophobic_to_aliphatic_hydrophobic  \n",
       "31055         positive_polar_to_aliphatic_hydrophobic  \n",
       "31056                  neutral_polar_to_neutral_polar  \n",
       "31057         negative_polar_to_aliphatic_hydrophobic  \n",
       "\n",
       "[31058 rows x 120 columns]"
      ]
     },
     "execution_count": 143,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinvar_filtered"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 144,
   "id": "b84d5ea5-3677-4530-8cc3-721f297aaa8d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "special_to_positive_polar                         2177\n",
       "aliphatic_hydrophobic_to_aliphatic_hydrophobic    2099\n",
       "aliphatic_hydrophobic_to_special                  1705\n",
       "positive_polar_to_special                         1658\n",
       "special_to_aliphatic_hydrophobic                  1605\n",
       "aliphatic_hydrophobic_to_neutral_polar            1571\n",
       "special_to_neutral_polar                          1538\n",
       "positive_polar_to_neutral_polar                   1398\n",
       "special_to_aromatic_hydrophobic                   1320\n",
       "special_to_negative_polar                         1212\n",
       "neutral_polar_to_positive_polar                   1195\n",
       "neutral_polar_to_aliphatic_hydrophobic            1027\n",
       "aromatic_hydrophobic_to_special                    894\n",
       "positive_polar_to_positive_polar                   870\n",
       "neutral_polar_to_special                           819\n",
       "negative_polar_to_positive_polar                   818\n",
       "aliphatic_hydrophobic_to_positive_polar            761\n",
       "positive_polar_to_aromatic_hydrophobic             671\n",
       "aromatic_hydrophobic_to_aliphatic_hydrophobic      603\n",
       "aliphatic_hydrophobic_to_aromatic_hydrophobic      602\n",
       "neutral_polar_to_neutral_polar                     599\n",
       "negative_polar_to_special                          573\n",
       "aliphatic_hydrophobic_to_negative_polar            545\n",
       "positive_polar_to_aliphatic_hydrophobic            542\n",
       "aromatic_hydrophobic_to_neutral_polar              539\n",
       "negative_polar_to_aliphatic_hydrophobic            508\n",
       "negative_polar_to_neutral_polar                    497\n",
       "special_to_special                                 491\n",
       "aromatic_hydrophobic_to_positive_polar             468\n",
       "neutral_polar_to_aromatic_hydrophobic              455\n",
       "negative_polar_to_negative_polar                   313\n",
       "positive_polar_to_negative_polar                   309\n",
       "negative_polar_to_aromatic_hydrophobic             285\n",
       "neutral_polar_to_negative_polar                    224\n",
       "aromatic_hydrophobic_to_negative_polar             104\n",
       "aromatic_hydrophobic_to_aromatic_hydrophobic        63\n",
       "Name: changes, dtype: int64"
      ]
     },
     "execution_count": 144,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinvar_filtered['changes'].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "13e96bdb-e965-4650-9ac7-38edaf24f449",
   "metadata": {},
   "source": [
    "### one-hot encoded and separation for columns: foldx & secondary structure (H,E,T,S,G,B,I) & is_catalytic & blosum62 (conservative & non-conservative)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 158,
   "id": "14df4650-11f1-46b4-9398-64e2b9910f5d",
   "metadata": {},
   "outputs": [],
   "source": [
    "clinvar_expanded = clinvar_filtered.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 159,
   "id": "c7cc846b-dc57-4539-8dd0-7b390ee4c47a",
   "metadata": {},
   "outputs": [],
   "source": [
    "one_hot = pd.get_dummies(clinvar_expanded['secondary_structure'], prefix='ss')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 160,
   "id": "30db6241-c45f-4b86-bbcb-51b7da1aa730",
   "metadata": {},
   "outputs": [],
   "source": [
    "clinvar_expanded = pd.concat([clinvar_expanded, one_hot], axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 165,
   "id": "31215838-3d3f-4b41-90de-03d6716d848e",
   "metadata": {},
   "outputs": [],
   "source": [
    "clinvar_expanded['is_catalytic'] = clinvar_expanded['is_catalytic'].astype(int)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 166,
   "id": "0327ea22-17d3-45c7-a579-2fd809f289f5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0        0\n",
       "1        0\n",
       "2        0\n",
       "3        0\n",
       "4        0\n",
       "5        0\n",
       "6        0\n",
       "        ..\n",
       "31051    0\n",
       "31052    0\n",
       "31053    0\n",
       "31054    0\n",
       "31055    0\n",
       "31056    0\n",
       "31057    0\n",
       "Name: is_catalytic, Length: 31058, dtype: int64"
      ]
     },
     "execution_count": 166,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinvar_expanded['is_catalytic']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 167,
   "id": "f34a76aa-a2fe-4b1e-8615-1c02d53ab2cd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    30827\n",
       "1      231\n",
       "Name: is_catalytic, dtype: int64"
      ]
     },
     "execution_count": 167,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinvar_expanded['is_catalytic'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 176,
   "id": "1f5840e9-45a8-4fc4-94c1-e515ecd0c61b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 176,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinvar_expanded['BLOSUM62'].isnull().sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 171,
   "id": "bccbcbe6-caaf-4024-98fe-529520da14b7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   BLOSUM62  conservative_changes  nonconservative_changes\n",
      "0        -2                   NaN                      2.0\n",
      "1         5                   5.0                      NaN\n",
      "2         0                   0.0                      NaN\n",
      "3        -1                   NaN                      1.0\n",
      "4        11                  11.0                      NaN\n",
      "5        -4                   NaN                      4.0\n",
      "6         3                   3.0                      NaN\n",
      "7        -3                   NaN                      3.0\n",
      "8         7                   7.0                      NaN\n"
     ]
    }
   ],
   "source": [
    "# Example dataframe\n",
    "df = pd.DataFrame({'BLOSUM62': [-2, 5, 0, -1, 11, -4, 3, -3, 7]})\n",
    "\n",
    "# Create new columns and assign NaN values\n",
    "df['conservative_changes'] = np.nan\n",
    "df['nonconservative_changes'] = np.nan\n",
    "\n",
    "# Set values based on conditions\n",
    "df.loc[df['BLOSUM62'].between(0, 11), 'conservative_changes'] = df['BLOSUM62']\n",
    "df.loc[df['BLOSUM62'].between(-4, -1), 'nonconservative_changes'] = df['BLOSUM62'].abs()\n",
    "\n",
    "print(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 177,
   "id": "e1369cc4-6e1e-4c80-aa30-cddf169f32ac",
   "metadata": {},
   "outputs": [],
   "source": [
    "clinvar_expanded['BLOSUM62_nonNeg'] = clinvar_expanded['BLOSUM62'].apply(lambda x: max(0, x + 4))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 180,
   "id": "03473c27-8395-4567-9daa-3dbf5469dbef",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['#CHROM',\n",
       " 'POS',\n",
       " 'ID',\n",
       " 'REF',\n",
       " 'ALT',\n",
       " 'AF_ESP',\n",
       " 'AF_EXAC',\n",
       " 'AF_TGP',\n",
       " 'ALLELEID',\n",
       " 'CLNDN',\n",
       " 'CLNDNINCL',\n",
       " 'CLNDISDB',\n",
       " 'CLNDISDBINCL',\n",
       " 'CLNHGVS',\n",
       " 'CLNREVSTAT',\n",
       " 'CLNSIG',\n",
       " 'CLNSIGCONF',\n",
       " 'CLNSIGINCL',\n",
       " 'CLNVC',\n",
       " 'CLNVCSO',\n",
       " 'CLNVI',\n",
       " 'DBVARID',\n",
       " 'MC',\n",
       " 'ORIGIN',\n",
       " 'RS',\n",
       " 'Allele',\n",
       " 'Consequence',\n",
       " 'IMPACT',\n",
       " 'SYMBOL',\n",
       " 'Gene',\n",
       " 'Feature_type',\n",
       " 'Feature',\n",
       " 'BIOTYPE',\n",
       " 'EXON',\n",
       " 'INTRON',\n",
       " 'HGVSc',\n",
       " 'HGVSp',\n",
       " 'cDNA_position',\n",
       " 'CDS_position',\n",
       " 'Protein_position',\n",
       " 'Amino_acids',\n",
       " 'Codons',\n",
       " 'Existing_variation',\n",
       " 'DISTANCE',\n",
       " 'STRAND',\n",
       " 'VARIANT_CLASS',\n",
       " 'SYMBOL_SOURCE',\n",
       " 'HGNC_ID',\n",
       " 'CANONICAL',\n",
       " 'MANE_SELECT',\n",
       " 'MANE_PLUS_CLINICAL',\n",
       " 'ENSP',\n",
       " 'SWISSPROT',\n",
       " 'TREMBL',\n",
       " 'UNIPARC',\n",
       " 'UNIPROT_ISOFORM',\n",
       " 'RefSeq',\n",
       " 'SIFT',\n",
       " 'PolyPhen',\n",
       " 'DOMAINS',\n",
       " 'CLIN_SIG',\n",
       " 'SOMATIC',\n",
       " 'PHENO',\n",
       " 'Conservation',\n",
       " 'BLOSUM62',\n",
       " 'GENEINFO_NAME',\n",
       " 'GENEINFO_ID',\n",
       " 'RefSeq_noversion',\n",
       " 'UniProt_IDs',\n",
       " 'PDB_path',\n",
       " 'Residue_position',\n",
       " 'Residue',\n",
       " 'DSSP_path',\n",
       " 'intra_contacts',\n",
       " 'is_catalytic',\n",
       " 'pLDDT',\n",
       " 'secondary_structure',\n",
       " 'accessibility',\n",
       " 'total energy',\n",
       " 'Backbone Hbond',\n",
       " 'Sidechain Hbond',\n",
       " 'Van der Waals',\n",
       " 'Electrostatics',\n",
       " 'Solvation Polar',\n",
       " 'Solvation Hydrophobic',\n",
       " 'Van der Waals clashes',\n",
       " 'entropy sidechain',\n",
       " 'entropy mainchain',\n",
       " 'sloop_entropy',\n",
       " 'mloop_entropy',\n",
       " 'cis_bond',\n",
       " 'torsional clash',\n",
       " 'backbone clash',\n",
       " 'helix dipole',\n",
       " 'water bridge',\n",
       " 'disulfide',\n",
       " 'electrostatic kon',\n",
       " 'partial covalent bonds',\n",
       " 'energy Ionisation',\n",
       " 'Entropy Complex',\n",
       " 'DiseaseClass',\n",
       " 'WT',\n",
       " 'MUT',\n",
       " 'aliphatic_assign_WT',\n",
       " 'aromatic_assign_WT',\n",
       " 'hydrophobic_assign_WT',\n",
       " 'positive_assign_WT',\n",
       " 'negative_assign_WT',\n",
       " 'neutral_assign_WT',\n",
       " 'polar_assign_WT',\n",
       " 'special_assign_WT',\n",
       " 'aliphatic_assign_MUT',\n",
       " 'aromatic_assign_MUT',\n",
       " 'hydrophobic_assign_MUT',\n",
       " 'positive_assign_MUT',\n",
       " 'negative_assign_MUT',\n",
       " 'neutral_assign_MUT',\n",
       " 'polar_assign_MUT',\n",
       " 'special_assign_MUT',\n",
       " 'changes',\n",
       " 'ss_B',\n",
       " 'ss_E',\n",
       " 'ss_G',\n",
       " 'ss_H',\n",
       " 'ss_I',\n",
       " 'ss_S',\n",
       " 'ss_T',\n",
       " 'BLOSUM62_nonNeg']"
      ]
     },
     "execution_count": 180,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinvar_expanded.columns.to_list()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 193,
   "id": "fad30e31-b7eb-44fa-a39f-7c7aa6aadb88",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2"
      ]
     },
     "execution_count": 193,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinvar_expanded['DOMAINS'].isnull().sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 194,
   "id": "056bf86e-5568-46f5-bae9-4c4afe0f3208",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "YES    30516\n",
       "Name: CANONICAL, dtype: int64"
      ]
     },
     "execution_count": 194,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinvar_expanded['CANONICAL'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 196,
   "id": "beb4bd63-c7ab-4634-9eac-84292a782dd1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "NM_001371246.1    185\n",
       "NM_014191.4       115\n",
       "NM_001080463.2     50\n",
       "NM_004992.4        43\n",
       "NM_001099404.2     38\n",
       "NM_003494.4        30\n",
       "NM_001159702.3     17\n",
       "NM_000828.5        12\n",
       "NM_001006657.2      9\n",
       "NM_030813.6         8\n",
       "NM_015359.6         6\n",
       "NM_001369387.1      6\n",
       "NM_017890.5         5\n",
       "NM_033056.4         5\n",
       "NM_133379.5         4\n",
       "NM_001365677.2      2\n",
       "NM_005888.4         2\n",
       "NM_001164507.2      2\n",
       "NM_005709.4         1\n",
       "NM_001329556.3      1\n",
       "NM_001394928.1      1\n",
       "Name: MANE_PLUS_CLINICAL, dtype: int64"
      ]
     },
     "execution_count": 196,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinvar_expanded['MANE_PLUS_CLINICAL'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 202,
   "id": "07c56955-08e6-4f3f-9b29-956790f493c4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "#CHROM                                                                    Y\n",
       "POS                                                                 2787515\n",
       "ID                                                                   492908\n",
       "REF                                                                       C\n",
       "ALT                                                                       A\n",
       "AF_ESP                                                                  NaN\n",
       "AF_EXAC                                                                 NaN\n",
       "AF_TGP                                                                  NaN\n",
       "ALLELEID                                                             485861\n",
       "CLNDN                                                  46,XY_sex_reversal_1\n",
       "CLNDNINCL                                                               NaN\n",
       "CLNDISDB                  MONDO:MONDO:0020712,MedGen:C2748896,OMIM:40004...\n",
       "CLNDISDBINCL                                                            NaN\n",
       "CLNHGVS                                           NC_000024.10:g.2787515C>A\n",
       "CLNREVSTAT                                   no_assertion_criteria_provided\n",
       "CLNSIG                                                           Pathogenic\n",
       "CLNSIGCONF                                                              NaN\n",
       "CLNSIGINCL                                                              NaN\n",
       "CLNVC                                             single_nucleotide_variant\n",
       "CLNVCSO                                                          SO:0001483\n",
       "CLNVI                                                   ClinGen:CA414941829\n",
       "DBVARID                                                                 NaN\n",
       "MC                                              SO:0001583|missense_variant\n",
       "ORIGIN                                                                    1\n",
       "RS                                                               1556370576\n",
       "Allele                                                                    A\n",
       "Consequence                                                missense_variant\n",
       "IMPACT                                                             MODERATE\n",
       "SYMBOL                                                                  SRY\n",
       "Gene                                                        ENSG00000184895\n",
       "Feature_type                                                     Transcript\n",
       "Feature                                                     ENST00000383070\n",
       "BIOTYPE                                                      protein_coding\n",
       "EXON                                                                    1/1\n",
       "INTRON                                                                  NaN\n",
       "HGVSc                                                                   NaN\n",
       "HGVSp                                                                   NaN\n",
       "cDNA_position                                                       168/828\n",
       "CDS_position                                                         89/615\n",
       "Protein_position                                                     30/204\n",
       "Amino_acids                                                             R/I\n",
       "Codons                                                              aGa/aTa\n",
       "Existing_variation                      rs1556370576&CM023171&COSV105326851\n",
       "DISTANCE                                                                NaN\n",
       "STRAND                                                                   -1\n",
       "VARIANT_CLASS                                                           SNV\n",
       "SYMBOL_SOURCE                                                          HGNC\n",
       "HGNC_ID                                                          HGNC:11311\n",
       "CANONICAL                                                               YES\n",
       "MANE_SELECT                                                     NM_003140.3\n",
       "MANE_PLUS_CLINICAL                                                      NaN\n",
       "ENSP                                                        ENSP00000372547\n",
       "SWISSPROT                                                        Q05066.216\n",
       "TREMBL                                                           A7WPU8.138\n",
       "UNIPARC                                                       UPI0000135F78\n",
       "UNIPROT_ISOFORM                                                         NaN\n",
       "RefSeq                                                          NM_003140.3\n",
       "SIFT                                                      deleterious(0.01)\n",
       "PolyPhen                                                                NaN\n",
       "DOMAINS                   AFDB-ENSP_mappings:AF-Q05066-F1.A&PIRSF:PIRSF0...\n",
       "CLIN_SIG                                                         pathogenic\n",
       "SOMATIC                                                               0&0&1\n",
       "PHENO                                                                 1&1&1\n",
       "Conservation                                                            NaN\n",
       "BLOSUM62                                                                 -3\n",
       "GENEINFO_NAME                                                           SRY\n",
       "GENEINFO_ID                                                            6736\n",
       "RefSeq_noversion                                                  NM_003140\n",
       "UniProt_IDs                                                [Q05066, A7WPU8]\n",
       "PDB_path                  /ibex/scratch/projects/c2102/databases/alphafo...\n",
       "Residue_position                                                         30\n",
       "Residue                                                                 Arg\n",
       "DSSP_path                 /ibex/scratch/projects/c2102/databases/dssp/al...\n",
       "intra_contacts                                                           []\n",
       "is_catalytic                                                              0\n",
       "pLDDT                                                             41.185997\n",
       "secondary_structure                                                     NaN\n",
       "accessibility                                                           242\n",
       "total energy                                                         1.4201\n",
       "Backbone Hbond                                                     0.698396\n",
       "Sidechain Hbond                                                    0.698396\n",
       "Van der Waals                                                     -0.024134\n",
       "Electrostatics                                                    -0.036773\n",
       "Solvation Polar                                                     0.09021\n",
       "Solvation Hydrophobic                                             -0.056088\n",
       "Van der Waals clashes                                             -0.004979\n",
       "entropy sidechain                                                 -0.681208\n",
       "entropy mainchain                                                  0.055582\n",
       "sloop_entropy                                                           0.0\n",
       "mloop_entropy                                                           0.0\n",
       "cis_bond                                                                0.0\n",
       "torsional clash                                                    0.680702\n",
       "backbone clash                                                          0.0\n",
       "helix dipole                                                            0.0\n",
       "water bridge                                                            0.0\n",
       "disulfide                                                               0.0\n",
       "electrostatic kon                                                       0.0\n",
       "partial covalent bonds                                                  0.0\n",
       "energy Ionisation                                                       0.0\n",
       "Entropy Complex                                                         0.0\n",
       "DiseaseClass                                                           None\n",
       "WT                                                                      Arg\n",
       "MUT                                                                     Ile\n",
       "aliphatic_assign_WT                                                       0\n",
       "aromatic_assign_WT                                                        0\n",
       "hydrophobic_assign_WT                                                     0\n",
       "positive_assign_WT                                                        1\n",
       "negative_assign_WT                                                        0\n",
       "neutral_assign_WT                                                         0\n",
       "polar_assign_WT                                                           1\n",
       "special_assign_WT                                                         0\n",
       "aliphatic_assign_MUT                                                      1\n",
       "aromatic_assign_MUT                                                       0\n",
       "hydrophobic_assign_MUT                                                    1\n",
       "positive_assign_MUT                                                       0\n",
       "negative_assign_MUT                                                       0\n",
       "neutral_assign_MUT                                                        0\n",
       "polar_assign_MUT                                                          0\n",
       "special_assign_MUT                                                        0\n",
       "changes                             positive_polar_to_aliphatic_hydrophobic\n",
       "ss_B                                                                      0\n",
       "ss_E                                                                      0\n",
       "ss_G                                                                      0\n",
       "ss_H                                                                      0\n",
       "ss_I                                                                      0\n",
       "ss_S                                                                      0\n",
       "ss_T                                                                      0\n",
       "BLOSUM62_nonNeg                                                           1\n",
       "Name: 31055, dtype: object"
      ]
     },
     "execution_count": 202,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinvar_expanded.loc[31055]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 189,
   "id": "9ad74889-077e-42f8-9b49-fb737cd08134",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "PosixPath('/ibex/scratch/projects/c2102/databases/alphafold/alphafold_mane/AF-P82279-F1-model_v4.pdb')"
      ]
     },
     "execution_count": 189,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinvar_expanded['PDB_path'].loc[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 181,
   "id": "0622abb6-ab4a-4ad4-8e62-a50259a8f501",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting ensembl-rest\n",
      "  Downloading ensembl_rest-0.3.3-py3-none-any.whl (34 kB)\n",
      "Requirement already satisfied: importlib-resources in /Users/penaguka/miniconda3/lib/python3.9/site-packages (from ensembl-rest) (5.2.0)\n",
      "Requirement already satisfied: requests in /Users/penaguka/miniconda3/lib/python3.9/site-packages (from ensembl-rest) (2.28.1)\n",
      "Requirement already satisfied: simplejson in /Users/penaguka/miniconda3/lib/python3.9/site-packages (from ensembl-rest) (3.19.1)\n",
      "Collecting intervaltree (from ensembl-rest)\n",
      "  Downloading intervaltree-3.1.0.tar.gz (32 kB)\n",
      "  Preparing metadata (setup.py) ... \u001b[?25ldone\n",
      "\u001b[?25hRequirement already satisfied: zipp>=3.1.0 in /Users/penaguka/miniconda3/lib/python3.9/site-packages (from importlib-resources->ensembl-rest) (3.11.0)\n",
      "Collecting sortedcontainers<3.0,>=2.0 (from intervaltree->ensembl-rest)\n",
      "  Downloading sortedcontainers-2.4.0-py2.py3-none-any.whl (29 kB)\n",
      "Requirement already satisfied: charset-normalizer<3,>=2 in /Users/penaguka/miniconda3/lib/python3.9/site-packages (from requests->ensembl-rest) (2.0.4)\n",
      "Requirement already satisfied: idna<4,>=2.5 in /Users/penaguka/miniconda3/lib/python3.9/site-packages (from requests->ensembl-rest) (3.4)\n",
      "Requirement already satisfied: urllib3<1.27,>=1.21.1 in /Users/penaguka/miniconda3/lib/python3.9/site-packages (from requests->ensembl-rest) (1.26.14)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in /Users/penaguka/miniconda3/lib/python3.9/site-packages (from requests->ensembl-rest) (2022.12.7)\n",
      "Building wheels for collected packages: intervaltree\n",
      "  Building wheel for intervaltree (setup.py) ... \u001b[?25ldone\n",
      "\u001b[?25h  Created wheel for intervaltree: filename=intervaltree-3.1.0-py2.py3-none-any.whl size=26098 sha256=539f8702b7784f392819d617557d0c807ae20821512310395085cc770f198e78\n",
      "  Stored in directory: /Users/penaguka/Library/Caches/pip/wheels/ab/fa/1b/75d9a713279796785711bd0bad8334aaace560c0bd28830c8c\n",
      "Successfully built intervaltree\n",
      "Installing collected packages: sortedcontainers, intervaltree, ensembl-rest\n",
      "Successfully installed ensembl-rest-0.3.3 intervaltree-3.1.0 sortedcontainers-2.4.0\n"
     ]
    }
   ],
   "source": [
    "!pip install ensembl-rest"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 204,
   "id": "6d6ac3e7-f156-4036-8f55-f5e56dfd2f26",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Protein Length: 204\n"
     ]
    }
   ],
   "source": [
    "import requests\n",
    "\n",
    "def get_protein_length(transcript_id):\n",
    "    if transcript_id in protein_length_cache:\n",
    "        # If the protein length is already cached, return it\n",
    "        return protein_length_cache[transcript_id]\n",
    "    \n",
    "    url = f\"https://rest.ensembl.org/sequence/id/{transcript_id}?type=protein\"\n",
    "    headers = {\"Content-Type\": \"application/json\"}\n",
    "\n",
    "    response = requests.get(url, headers=headers)\n",
    "\n",
    "    if response.ok:\n",
    "        protein_sequence = response.json()['seq']\n",
    "        protein_length = len(protein_sequence)\n",
    "        return protein_length\n",
    "    else:\n",
    "        print(f\"Error retrieving protein length of {transcript_id}.\")\n",
    "        # Handle error if transcript ID is invalid or protein sequence is not available\n",
    "        return None\n",
    "\n",
    "# Example usage\n",
    "transcript_id = \"ENST00000383070\"  # Replace with your transcript ID\n",
    "length = get_protein_length(transcript_id)\n",
    "\n",
    "if length is not None:\n",
    "    print(f\"Protein Length: {length}\")\n",
    "else:\n",
    "    print(\"Error retrieving protein length.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 205,
   "id": "5202706e-60d6-47f8-9a83-4bdca2e010c0",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "ename": "ConnectionError",
     "evalue": "HTTPSConnectionPool(host='rest.ensembl.org', port=443): Max retries exceeded with url: /sequence/id/ENST00000296589?type=protein (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x7fe5d7f7f520>: Failed to establish a new connection: [Errno 8] nodename nor servname provided, or not known'))",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mgaierror\u001b[0m                                  Traceback (most recent call last)",
      "File \u001b[0;32m~/miniconda3/lib/python3.9/site-packages/urllib3/connection.py:174\u001b[0m, in \u001b[0;36mHTTPConnection._new_conn\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    173\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m--> 174\u001b[0m     conn \u001b[38;5;241m=\u001b[39m \u001b[43mconnection\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcreate_connection\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    175\u001b[0m \u001b[43m        \u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_dns_host\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mport\u001b[49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mtimeout\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mextra_kw\u001b[49m\n\u001b[1;32m    176\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    178\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m SocketTimeout:\n",
      "File \u001b[0;32m~/miniconda3/lib/python3.9/site-packages/urllib3/util/connection.py:72\u001b[0m, in \u001b[0;36mcreate_connection\u001b[0;34m(address, timeout, source_address, socket_options)\u001b[0m\n\u001b[1;32m     68\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m six\u001b[38;5;241m.\u001b[39mraise_from(\n\u001b[1;32m     69\u001b[0m         LocationParseError(\u001b[38;5;124mu\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;132;01m%s\u001b[39;00m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m, label empty or too long\u001b[39m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;241m%\u001b[39m host), \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[1;32m     70\u001b[0m     )\n\u001b[0;32m---> 72\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m res \u001b[38;5;129;01min\u001b[39;00m \u001b[43msocket\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mgetaddrinfo\u001b[49m\u001b[43m(\u001b[49m\u001b[43mhost\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mport\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mfamily\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43msocket\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mSOCK_STREAM\u001b[49m\u001b[43m)\u001b[49m:\n\u001b[1;32m     73\u001b[0m     af, socktype, proto, canonname, sa \u001b[38;5;241m=\u001b[39m res\n",
      "File \u001b[0;32m~/miniconda3/lib/python3.9/socket.py:954\u001b[0m, in \u001b[0;36mgetaddrinfo\u001b[0;34m(host, port, family, type, proto, flags)\u001b[0m\n\u001b[1;32m    953\u001b[0m addrlist \u001b[38;5;241m=\u001b[39m []\n\u001b[0;32m--> 954\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m res \u001b[38;5;129;01min\u001b[39;00m \u001b[43m_socket\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mgetaddrinfo\u001b[49m\u001b[43m(\u001b[49m\u001b[43mhost\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mport\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mfamily\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mtype\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mproto\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mflags\u001b[49m\u001b[43m)\u001b[49m:\n\u001b[1;32m    955\u001b[0m     af, socktype, proto, canonname, sa \u001b[38;5;241m=\u001b[39m res\n",
      "\u001b[0;31mgaierror\u001b[0m: [Errno 8] nodename nor servname provided, or not known",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[0;31mNewConnectionError\u001b[0m                        Traceback (most recent call last)",
      "File \u001b[0;32m~/miniconda3/lib/python3.9/site-packages/urllib3/connectionpool.py:703\u001b[0m, in \u001b[0;36mHTTPConnectionPool.urlopen\u001b[0;34m(self, method, url, body, headers, retries, redirect, assert_same_host, timeout, pool_timeout, release_conn, chunked, body_pos, **response_kw)\u001b[0m\n\u001b[1;32m    702\u001b[0m \u001b[38;5;66;03m# Make the request on the httplib connection object.\u001b[39;00m\n\u001b[0;32m--> 703\u001b[0m httplib_response \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_make_request\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    704\u001b[0m \u001b[43m    \u001b[49m\u001b[43mconn\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    705\u001b[0m \u001b[43m    \u001b[49m\u001b[43mmethod\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    706\u001b[0m \u001b[43m    \u001b[49m\u001b[43murl\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    707\u001b[0m \u001b[43m    \u001b[49m\u001b[43mtimeout\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mtimeout_obj\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    708\u001b[0m \u001b[43m    \u001b[49m\u001b[43mbody\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mbody\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    709\u001b[0m \u001b[43m    \u001b[49m\u001b[43mheaders\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mheaders\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    710\u001b[0m \u001b[43m    \u001b[49m\u001b[43mchunked\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mchunked\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    711\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    713\u001b[0m \u001b[38;5;66;03m# If we're going to release the connection in ``finally:``, then\u001b[39;00m\n\u001b[1;32m    714\u001b[0m \u001b[38;5;66;03m# the response doesn't need to know about the connection. Otherwise\u001b[39;00m\n\u001b[1;32m    715\u001b[0m \u001b[38;5;66;03m# it will also try to release it and we'll have a double-release\u001b[39;00m\n\u001b[1;32m    716\u001b[0m \u001b[38;5;66;03m# mess.\u001b[39;00m\n",
      "File \u001b[0;32m~/miniconda3/lib/python3.9/site-packages/urllib3/connectionpool.py:386\u001b[0m, in \u001b[0;36mHTTPConnectionPool._make_request\u001b[0;34m(self, conn, method, url, timeout, chunked, **httplib_request_kw)\u001b[0m\n\u001b[1;32m    385\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m--> 386\u001b[0m     \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_validate_conn\u001b[49m\u001b[43m(\u001b[49m\u001b[43mconn\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    387\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m (SocketTimeout, BaseSSLError) \u001b[38;5;28;01mas\u001b[39;00m e:\n\u001b[1;32m    388\u001b[0m     \u001b[38;5;66;03m# Py2 raises this as a BaseSSLError, Py3 raises it as socket timeout.\u001b[39;00m\n",
      "File \u001b[0;32m~/miniconda3/lib/python3.9/site-packages/urllib3/connectionpool.py:1042\u001b[0m, in \u001b[0;36mHTTPSConnectionPool._validate_conn\u001b[0;34m(self, conn)\u001b[0m\n\u001b[1;32m   1041\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28mgetattr\u001b[39m(conn, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124msock\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;28;01mNone\u001b[39;00m):  \u001b[38;5;66;03m# AppEngine might not have  `.sock`\u001b[39;00m\n\u001b[0;32m-> 1042\u001b[0m     \u001b[43mconn\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mconnect\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1044\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m conn\u001b[38;5;241m.\u001b[39mis_verified:\n",
      "File \u001b[0;32m~/miniconda3/lib/python3.9/site-packages/urllib3/connection.py:358\u001b[0m, in \u001b[0;36mHTTPSConnection.connect\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    356\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mconnect\u001b[39m(\u001b[38;5;28mself\u001b[39m):\n\u001b[1;32m    357\u001b[0m     \u001b[38;5;66;03m# Add certificate verification\u001b[39;00m\n\u001b[0;32m--> 358\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39msock \u001b[38;5;241m=\u001b[39m conn \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_new_conn\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    359\u001b[0m     hostname \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mhost\n",
      "File \u001b[0;32m~/miniconda3/lib/python3.9/site-packages/urllib3/connection.py:186\u001b[0m, in \u001b[0;36mHTTPConnection._new_conn\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    185\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m SocketError \u001b[38;5;28;01mas\u001b[39;00m e:\n\u001b[0;32m--> 186\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m NewConnectionError(\n\u001b[1;32m    187\u001b[0m         \u001b[38;5;28mself\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mFailed to establish a new connection: \u001b[39m\u001b[38;5;132;01m%s\u001b[39;00m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;241m%\u001b[39m e\n\u001b[1;32m    188\u001b[0m     )\n\u001b[1;32m    190\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m conn\n",
      "\u001b[0;31mNewConnectionError\u001b[0m: <urllib3.connection.HTTPSConnection object at 0x7fe5d7f7f520>: Failed to establish a new connection: [Errno 8] nodename nor servname provided, or not known",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[0;31mMaxRetryError\u001b[0m                             Traceback (most recent call last)",
      "File \u001b[0;32m~/miniconda3/lib/python3.9/site-packages/requests/adapters.py:489\u001b[0m, in \u001b[0;36mHTTPAdapter.send\u001b[0;34m(self, request, stream, timeout, verify, cert, proxies)\u001b[0m\n\u001b[1;32m    488\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m chunked:\n\u001b[0;32m--> 489\u001b[0m     resp \u001b[38;5;241m=\u001b[39m \u001b[43mconn\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43murlopen\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    490\u001b[0m \u001b[43m        \u001b[49m\u001b[43mmethod\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mrequest\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mmethod\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    491\u001b[0m \u001b[43m        \u001b[49m\u001b[43murl\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43murl\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    492\u001b[0m \u001b[43m        \u001b[49m\u001b[43mbody\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mrequest\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mbody\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    493\u001b[0m \u001b[43m        \u001b[49m\u001b[43mheaders\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mrequest\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mheaders\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    494\u001b[0m \u001b[43m        \u001b[49m\u001b[43mredirect\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[1;32m    495\u001b[0m \u001b[43m        \u001b[49m\u001b[43massert_same_host\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[1;32m    496\u001b[0m \u001b[43m        \u001b[49m\u001b[43mpreload_content\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[1;32m    497\u001b[0m \u001b[43m        \u001b[49m\u001b[43mdecode_content\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[1;32m    498\u001b[0m \u001b[43m        \u001b[49m\u001b[43mretries\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mmax_retries\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    499\u001b[0m \u001b[43m        \u001b[49m\u001b[43mtimeout\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mtimeout\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    500\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    502\u001b[0m \u001b[38;5;66;03m# Send the request.\u001b[39;00m\n\u001b[1;32m    503\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n",
      "File \u001b[0;32m~/miniconda3/lib/python3.9/site-packages/urllib3/connectionpool.py:787\u001b[0m, in \u001b[0;36mHTTPConnectionPool.urlopen\u001b[0;34m(self, method, url, body, headers, retries, redirect, assert_same_host, timeout, pool_timeout, release_conn, chunked, body_pos, **response_kw)\u001b[0m\n\u001b[1;32m    785\u001b[0m     e \u001b[38;5;241m=\u001b[39m ProtocolError(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mConnection aborted.\u001b[39m\u001b[38;5;124m\"\u001b[39m, e)\n\u001b[0;32m--> 787\u001b[0m retries \u001b[38;5;241m=\u001b[39m \u001b[43mretries\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mincrement\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    788\u001b[0m \u001b[43m    \u001b[49m\u001b[43mmethod\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43murl\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43merror\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43me\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43m_pool\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43m_stacktrace\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43msys\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mexc_info\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;241;43m2\u001b[39;49m\u001b[43m]\u001b[49m\n\u001b[1;32m    789\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    790\u001b[0m retries\u001b[38;5;241m.\u001b[39msleep()\n",
      "File \u001b[0;32m~/miniconda3/lib/python3.9/site-packages/urllib3/util/retry.py:592\u001b[0m, in \u001b[0;36mRetry.increment\u001b[0;34m(self, method, url, response, error, _pool, _stacktrace)\u001b[0m\n\u001b[1;32m    591\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m new_retry\u001b[38;5;241m.\u001b[39mis_exhausted():\n\u001b[0;32m--> 592\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m MaxRetryError(_pool, url, error \u001b[38;5;129;01mor\u001b[39;00m ResponseError(cause))\n\u001b[1;32m    594\u001b[0m log\u001b[38;5;241m.\u001b[39mdebug(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mIncremented Retry for (url=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;132;01m%s\u001b[39;00m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m): \u001b[39m\u001b[38;5;132;01m%r\u001b[39;00m\u001b[38;5;124m\"\u001b[39m, url, new_retry)\n",
      "\u001b[0;31mMaxRetryError\u001b[0m: HTTPSConnectionPool(host='rest.ensembl.org', port=443): Max retries exceeded with url: /sequence/id/ENST00000296589?type=protein (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x7fe5d7f7f520>: Failed to establish a new connection: [Errno 8] nodename nor servname provided, or not known'))",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[0;31mConnectionError\u001b[0m                           Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[205], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m clinvar_expanded[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mprot_length\u001b[39m\u001b[38;5;124m'\u001b[39m] \u001b[38;5;241m=\u001b[39m \u001b[43mclinvar_expanded\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mFeature\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m]\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mapply\u001b[49m\u001b[43m(\u001b[49m\u001b[43mget_protein_length\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniconda3/lib/python3.9/site-packages/pandas/core/series.py:4771\u001b[0m, in \u001b[0;36mSeries.apply\u001b[0;34m(self, func, convert_dtype, args, **kwargs)\u001b[0m\n\u001b[1;32m   4661\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mapply\u001b[39m(\n\u001b[1;32m   4662\u001b[0m     \u001b[38;5;28mself\u001b[39m,\n\u001b[1;32m   4663\u001b[0m     func: AggFuncType,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m   4666\u001b[0m     \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs,\n\u001b[1;32m   4667\u001b[0m ) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m DataFrame \u001b[38;5;241m|\u001b[39m Series:\n\u001b[1;32m   4668\u001b[0m     \u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m   4669\u001b[0m \u001b[38;5;124;03m    Invoke function on values of Series.\u001b[39;00m\n\u001b[1;32m   4670\u001b[0m \n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m   4769\u001b[0m \u001b[38;5;124;03m    dtype: float64\u001b[39;00m\n\u001b[1;32m   4770\u001b[0m \u001b[38;5;124;03m    \"\"\"\u001b[39;00m\n\u001b[0;32m-> 4771\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mSeriesApply\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mfunc\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mconvert_dtype\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mapply\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniconda3/lib/python3.9/site-packages/pandas/core/apply.py:1123\u001b[0m, in \u001b[0;36mSeriesApply.apply\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m   1120\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mapply_str()\n\u001b[1;32m   1122\u001b[0m \u001b[38;5;66;03m# self.f is Callable\u001b[39;00m\n\u001b[0;32m-> 1123\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mapply_standard\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniconda3/lib/python3.9/site-packages/pandas/core/apply.py:1174\u001b[0m, in \u001b[0;36mSeriesApply.apply_standard\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m   1172\u001b[0m     \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m   1173\u001b[0m         values \u001b[38;5;241m=\u001b[39m obj\u001b[38;5;241m.\u001b[39mastype(\u001b[38;5;28mobject\u001b[39m)\u001b[38;5;241m.\u001b[39m_values\n\u001b[0;32m-> 1174\u001b[0m         mapped \u001b[38;5;241m=\u001b[39m \u001b[43mlib\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mmap_infer\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m   1175\u001b[0m \u001b[43m            \u001b[49m\u001b[43mvalues\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1176\u001b[0m \u001b[43m            \u001b[49m\u001b[43mf\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1177\u001b[0m \u001b[43m            \u001b[49m\u001b[43mconvert\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mconvert_dtype\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1178\u001b[0m \u001b[43m        \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1180\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mlen\u001b[39m(mapped) \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(mapped[\u001b[38;5;241m0\u001b[39m], ABCSeries):\n\u001b[1;32m   1181\u001b[0m     \u001b[38;5;66;03m# GH#43986 Need to do list(mapped) in order to get treated as nested\u001b[39;00m\n\u001b[1;32m   1182\u001b[0m     \u001b[38;5;66;03m#  See also GH#25959 regarding EA support\u001b[39;00m\n\u001b[1;32m   1183\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m obj\u001b[38;5;241m.\u001b[39m_constructor_expanddim(\u001b[38;5;28mlist\u001b[39m(mapped), index\u001b[38;5;241m=\u001b[39mobj\u001b[38;5;241m.\u001b[39mindex)\n",
      "File \u001b[0;32m~/miniconda3/lib/python3.9/site-packages/pandas/_libs/lib.pyx:2924\u001b[0m, in \u001b[0;36mpandas._libs.lib.map_infer\u001b[0;34m()\u001b[0m\n",
      "Cell \u001b[0;32mIn[204], line 7\u001b[0m, in \u001b[0;36mget_protein_length\u001b[0;34m(transcript_id)\u001b[0m\n\u001b[1;32m      4\u001b[0m url \u001b[38;5;241m=\u001b[39m \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mhttps://rest.ensembl.org/sequence/id/\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mtranscript_id\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m?type=protein\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m      5\u001b[0m headers \u001b[38;5;241m=\u001b[39m {\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mContent-Type\u001b[39m\u001b[38;5;124m\"\u001b[39m: \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mapplication/json\u001b[39m\u001b[38;5;124m\"\u001b[39m}\n\u001b[0;32m----> 7\u001b[0m response \u001b[38;5;241m=\u001b[39m \u001b[43mrequests\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[43murl\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mheaders\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mheaders\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m      9\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m response\u001b[38;5;241m.\u001b[39mok:\n\u001b[1;32m     10\u001b[0m     protein_sequence \u001b[38;5;241m=\u001b[39m response\u001b[38;5;241m.\u001b[39mjson()[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mseq\u001b[39m\u001b[38;5;124m'\u001b[39m]\n",
      "File \u001b[0;32m~/miniconda3/lib/python3.9/site-packages/requests/api.py:73\u001b[0m, in \u001b[0;36mget\u001b[0;34m(url, params, **kwargs)\u001b[0m\n\u001b[1;32m     62\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mget\u001b[39m(url, params\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mNone\u001b[39;00m, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs):\n\u001b[1;32m     63\u001b[0m     \u001b[38;5;124mr\u001b[39m\u001b[38;5;124;03m\"\"\"Sends a GET request.\u001b[39;00m\n\u001b[1;32m     64\u001b[0m \n\u001b[1;32m     65\u001b[0m \u001b[38;5;124;03m    :param url: URL for the new :class:`Request` object.\u001b[39;00m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m     70\u001b[0m \u001b[38;5;124;03m    :rtype: requests.Response\u001b[39;00m\n\u001b[1;32m     71\u001b[0m \u001b[38;5;124;03m    \"\"\"\u001b[39;00m\n\u001b[0;32m---> 73\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mrequest\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mget\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43murl\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mparams\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mparams\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniconda3/lib/python3.9/site-packages/requests/api.py:59\u001b[0m, in \u001b[0;36mrequest\u001b[0;34m(method, url, **kwargs)\u001b[0m\n\u001b[1;32m     55\u001b[0m \u001b[38;5;66;03m# By using the 'with' statement we are sure the session is closed, thus we\u001b[39;00m\n\u001b[1;32m     56\u001b[0m \u001b[38;5;66;03m# avoid leaving sockets open which can trigger a ResourceWarning in some\u001b[39;00m\n\u001b[1;32m     57\u001b[0m \u001b[38;5;66;03m# cases, and look like a memory leak in others.\u001b[39;00m\n\u001b[1;32m     58\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m sessions\u001b[38;5;241m.\u001b[39mSession() \u001b[38;5;28;01mas\u001b[39;00m session:\n\u001b[0;32m---> 59\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43msession\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrequest\u001b[49m\u001b[43m(\u001b[49m\u001b[43mmethod\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mmethod\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43murl\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43murl\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniconda3/lib/python3.9/site-packages/requests/sessions.py:587\u001b[0m, in \u001b[0;36mSession.request\u001b[0;34m(self, method, url, params, data, headers, cookies, files, auth, timeout, allow_redirects, proxies, hooks, stream, verify, cert, json)\u001b[0m\n\u001b[1;32m    582\u001b[0m send_kwargs \u001b[38;5;241m=\u001b[39m {\n\u001b[1;32m    583\u001b[0m     \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mtimeout\u001b[39m\u001b[38;5;124m\"\u001b[39m: timeout,\n\u001b[1;32m    584\u001b[0m     \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mallow_redirects\u001b[39m\u001b[38;5;124m\"\u001b[39m: allow_redirects,\n\u001b[1;32m    585\u001b[0m }\n\u001b[1;32m    586\u001b[0m send_kwargs\u001b[38;5;241m.\u001b[39mupdate(settings)\n\u001b[0;32m--> 587\u001b[0m resp \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msend\u001b[49m\u001b[43m(\u001b[49m\u001b[43mprep\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43msend_kwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    589\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m resp\n",
      "File \u001b[0;32m~/miniconda3/lib/python3.9/site-packages/requests/sessions.py:701\u001b[0m, in \u001b[0;36mSession.send\u001b[0;34m(self, request, **kwargs)\u001b[0m\n\u001b[1;32m    698\u001b[0m start \u001b[38;5;241m=\u001b[39m preferred_clock()\n\u001b[1;32m    700\u001b[0m \u001b[38;5;66;03m# Send the request\u001b[39;00m\n\u001b[0;32m--> 701\u001b[0m r \u001b[38;5;241m=\u001b[39m \u001b[43madapter\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msend\u001b[49m\u001b[43m(\u001b[49m\u001b[43mrequest\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    703\u001b[0m \u001b[38;5;66;03m# Total elapsed time of the request (approximately)\u001b[39;00m\n\u001b[1;32m    704\u001b[0m elapsed \u001b[38;5;241m=\u001b[39m preferred_clock() \u001b[38;5;241m-\u001b[39m start\n",
      "File \u001b[0;32m~/miniconda3/lib/python3.9/site-packages/requests/adapters.py:565\u001b[0m, in \u001b[0;36mHTTPAdapter.send\u001b[0;34m(self, request, stream, timeout, verify, cert, proxies)\u001b[0m\n\u001b[1;32m    561\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(e\u001b[38;5;241m.\u001b[39mreason, _SSLError):\n\u001b[1;32m    562\u001b[0m         \u001b[38;5;66;03m# This branch is for urllib3 v1.22 and later.\u001b[39;00m\n\u001b[1;32m    563\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m SSLError(e, request\u001b[38;5;241m=\u001b[39mrequest)\n\u001b[0;32m--> 565\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mConnectionError\u001b[39;00m(e, request\u001b[38;5;241m=\u001b[39mrequest)\n\u001b[1;32m    567\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m ClosedPoolError \u001b[38;5;28;01mas\u001b[39;00m e:\n\u001b[1;32m    568\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mConnectionError\u001b[39;00m(e, request\u001b[38;5;241m=\u001b[39mrequest)\n",
      "\u001b[0;31mConnectionError\u001b[0m: HTTPSConnectionPool(host='rest.ensembl.org', port=443): Max retries exceeded with url: /sequence/id/ENST00000296589?type=protein (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x7fe5d7f7f520>: Failed to establish a new connection: [Errno 8] nodename nor servname provided, or not known'))"
     ]
    }
   ],
   "source": [
    "clinvar_expanded['prot_length'] = clinvar_expanded['Feature'].apply(get_protein_length)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9fc279d7-21ee-4b8f-a133-d7f852c25e27",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "### gnomAD variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "388d690f-6d2c-42ab-88d6-fd4763b4e831",
   "metadata": {},
   "outputs": [],
   "source": [
    "gnomad_variants = pd.read_pickle('variants/gnomad_variants.pkl')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "52997a33-b7ea-42c6-8eb3-c40936c54d6a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><div id=85b8654b-2aa5-4364-9f24-02b1f2107761 style=\"display:none; background-color:#9D6CFF; color:white; width:200px; height:30px; padding-left:5px; border-radius:4px; flex-direction:row; justify-content:space-around; align-items:center;\" onmouseover=\"this.style.backgroundColor='#BA9BF8'\" onmouseout=\"this.style.backgroundColor='#9D6CFF'\" onclick=\"window.commands?.execute('create-mitosheet-from-dataframe-output');\">See Full Dataframe in Mito</div> <script> if (window.commands?.hasCommand('create-mitosheet-from-dataframe-output')) document.getElementById('85b8654b-2aa5-4364-9f24-02b1f2107761').style.display = 'flex' </script> <table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>AC</th>\n",
       "      <th>AF</th>\n",
       "      <th>AN</th>\n",
       "      <th>QD</th>\n",
       "      <th>variant_type</th>\n",
       "      <th>...</th>\n",
       "      <th>cis_bond</th>\n",
       "      <th>torsional clash</th>\n",
       "      <th>backbone clash</th>\n",
       "      <th>helix dipole</th>\n",
       "      <th>water bridge</th>\n",
       "      <th>disulfide</th>\n",
       "      <th>electrostatic kon</th>\n",
       "      <th>partial covalent bonds</th>\n",
       "      <th>energy Ionisation</th>\n",
       "      <th>Entropy Complex</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>11706367</td>\n",
       "      <td>rs11121804</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>39444</td>\n",
       "      <td>0.158732</td>\n",
       "      <td>248494</td>\n",
       "      <td>13.47</td>\n",
       "      <td>mixed</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.018940</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>11790870</td>\n",
       "      <td>rs2274976</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>13949</td>\n",
       "      <td>0.0555184</td>\n",
       "      <td>251250</td>\n",
       "      <td>14.18</td>\n",
       "      <td>snv</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.165854</td>\n",
       "      <td>-0.304761</td>\n",
       "      <td>3.235800e-01</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>11824498</td>\n",
       "      <td>rs198400</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>250947</td>\n",
       "      <td>0.999399</td>\n",
       "      <td>251098</td>\n",
       "      <td>33.49</td>\n",
       "      <td>snv</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.140793</td>\n",
       "      <td>-1.049150</td>\n",
       "      <td>-7.725940e-03</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>11847591</td>\n",
       "      <td>rs5063</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>14048</td>\n",
       "      <td>0.0558586</td>\n",
       "      <td>251492</td>\n",
       "      <td>13.96</td>\n",
       "      <td>snv</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.104118</td>\n",
       "      <td>0.236702</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>11949899</td>\n",
       "      <td>rs7551175</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>45490</td>\n",
       "      <td>0.180978</td>\n",
       "      <td>251356</td>\n",
       "      <td>14.94</td>\n",
       "      <td>snv</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.006133</td>\n",
       "      <td>0.028216</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15936</th>\n",
       "      <td>chrX</td>\n",
       "      <td>72140250</td>\n",
       "      <td>rs7061150</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>12196</td>\n",
       "      <td>0.0674621</td>\n",
       "      <td>180783</td>\n",
       "      <td>19.16</td>\n",
       "      <td>snv</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.031276</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15937</th>\n",
       "      <td>chrY</td>\n",
       "      <td>6246522</td>\n",
       "      <td>rs201475959</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>1025</td>\n",
       "      <td>0.0780952</td>\n",
       "      <td>13125</td>\n",
       "      <td>30.14</td>\n",
       "      <td>snv</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.000050</td>\n",
       "      <td>-0.708370</td>\n",
       "      <td>-4.440890e-16</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15938</th>\n",
       "      <td>chrY</td>\n",
       "      <td>12720687</td>\n",
       "      <td>rs7067496</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>6033</td>\n",
       "      <td>0.0894759</td>\n",
       "      <td>67426</td>\n",
       "      <td>32.95</td>\n",
       "      <td>snv</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.002779</td>\n",
       "      <td>-0.131799</td>\n",
       "      <td>3.340370e-01</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-8.881780e-16</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15939</th>\n",
       "      <td>chrY</td>\n",
       "      <td>22180292</td>\n",
       "      <td>rs1478920645</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "      <td>0.142857</td>\n",
       "      <td>7</td>\n",
       "      <td>15.53</td>\n",
       "      <td>snv</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.002249</td>\n",
       "      <td>0.067966</td>\n",
       "      <td>-2.492200e-01</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.996460e-03</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15940</th>\n",
       "      <td>chrY</td>\n",
       "      <td>22406127</td>\n",
       "      <td>rs1195699969</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>9</td>\n",
       "      <td>1</td>\n",
       "      <td>9</td>\n",
       "      <td>24.33</td>\n",
       "      <td>snv</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.002249</td>\n",
       "      <td>0.067966</td>\n",
       "      <td>-2.492200e-01</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.996460e-03</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table></div>"
      ],
      "text/plain": [
       "      #CHROM       POS            ID REF ALT      AC         AF      AN  \\\n",
       "0       chr1  11706367    rs11121804   C   T   39444   0.158732  248494   \n",
       "1       chr1  11790870     rs2274976   C   T   13949  0.0555184  251250   \n",
       "2       chr1  11824498      rs198400   A   G  250947   0.999399  251098   \n",
       "3       chr1  11847591        rs5063   C   T   14048  0.0558586  251492   \n",
       "4       chr1  11949899     rs7551175   G   A   45490   0.180978  251356   \n",
       "...      ...       ...           ...  ..  ..     ...        ...     ...   \n",
       "15936   chrX  72140250     rs7061150   C   T   12196  0.0674621  180783   \n",
       "15937   chrY   6246522   rs201475959   A   G    1025  0.0780952   13125   \n",
       "15938   chrY  12720687     rs7067496   G   T    6033  0.0894759   67426   \n",
       "15939   chrY  22180292  rs1478920645   G   T       1   0.142857       7   \n",
       "15940   chrY  22406127  rs1195699969   C   A       9          1       9   \n",
       "\n",
       "          QD variant_type  ... cis_bond torsional clash backbone clash  \\\n",
       "0      13.47        mixed  ...      0.0       -0.018940       0.000000   \n",
       "1      14.18          snv  ...      0.0       -0.165854      -0.304761   \n",
       "2      33.49          snv  ...      0.0       -0.140793      -1.049150   \n",
       "3      13.96          snv  ...      0.0       -0.104118       0.236702   \n",
       "4      14.94          snv  ...      0.0        0.006133       0.028216   \n",
       "...      ...          ...  ...      ...             ...            ...   \n",
       "15936  19.16          snv  ...      0.0        0.031276       0.000000   \n",
       "15937  30.14          snv  ...      0.0       -0.000050      -0.708370   \n",
       "15938  32.95          snv  ...      0.0       -0.002779      -0.131799   \n",
       "15939  15.53          snv  ...      0.0        0.002249       0.067966   \n",
       "15940  24.33          snv  ...      0.0        0.002249       0.067966   \n",
       "\n",
       "       helix dipole water bridge disulfide electrostatic kon  \\\n",
       "0      0.000000e+00          0.0       0.0               0.0   \n",
       "1      3.235800e-01          0.0       0.0               0.0   \n",
       "2     -7.725940e-03          0.0       0.0               0.0   \n",
       "3      0.000000e+00          0.0       0.0               0.0   \n",
       "4      0.000000e+00          0.0       0.0               0.0   \n",
       "...             ...          ...       ...               ...   \n",
       "15936  0.000000e+00          0.0       0.0               0.0   \n",
       "15937 -4.440890e-16          0.0       0.0               0.0   \n",
       "15938  3.340370e-01          0.0       0.0               0.0   \n",
       "15939 -2.492200e-01          0.0       0.0               0.0   \n",
       "15940 -2.492200e-01          0.0       0.0               0.0   \n",
       "\n",
       "      partial covalent bonds energy Ionisation Entropy Complex  \n",
       "0                        0.0      0.000000e+00             0.0  \n",
       "1                        0.0      0.000000e+00             0.0  \n",
       "2                        0.0      0.000000e+00             0.0  \n",
       "3                        0.0      0.000000e+00             0.0  \n",
       "4                        0.0      0.000000e+00             0.0  \n",
       "...                      ...               ...             ...  \n",
       "15936                    0.0      0.000000e+00             0.0  \n",
       "15937                    0.0      0.000000e+00             0.0  \n",
       "15938                    0.0     -8.881780e-16             0.0  \n",
       "15939                    0.0     -1.996460e-03             0.0  \n",
       "15940                    0.0     -1.996460e-03             0.0  \n",
       "\n",
       "[15941 rows x 85 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gnomad_variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "93d246ed-cb28-4b19-b582-e7e26b248079",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['#CHROM',\n",
       " 'POS',\n",
       " 'ID',\n",
       " 'REF',\n",
       " 'ALT',\n",
       " 'AC',\n",
       " 'AF',\n",
       " 'AN',\n",
       " 'QD',\n",
       " 'variant_type',\n",
       " 'vep',\n",
       " 'flags',\n",
       " 'Allele',\n",
       " 'Consequence',\n",
       " 'IMPACT',\n",
       " 'SYMBOL',\n",
       " 'Gene',\n",
       " 'Feature_type',\n",
       " 'Feature',\n",
       " 'BIOTYPE',\n",
       " 'EXON',\n",
       " 'INTRON',\n",
       " 'HGVSc',\n",
       " 'HGVSp',\n",
       " 'cDNA_position',\n",
       " 'CDS_position',\n",
       " 'Protein_position',\n",
       " 'Amino_acids',\n",
       " 'Codons',\n",
       " 'Existing_variation',\n",
       " 'DISTANCE',\n",
       " 'STRAND',\n",
       " 'VARIANT_CLASS',\n",
       " 'SYMBOL_SOURCE',\n",
       " 'HGNC_ID',\n",
       " 'CANONICAL',\n",
       " 'MANE_SELECT',\n",
       " 'MANE_PLUS_CLINICAL',\n",
       " 'ENSP',\n",
       " 'SWISSPROT',\n",
       " 'TREMBL',\n",
       " 'UNIPARC',\n",
       " 'UNIPROT_ISOFORM',\n",
       " 'RefSeq',\n",
       " 'SIFT',\n",
       " 'PolyPhen',\n",
       " 'DOMAINS',\n",
       " 'CLIN_SIG',\n",
       " 'SOMATIC',\n",
       " 'PHENO',\n",
       " 'Conservation',\n",
       " 'BLOSUM62',\n",
       " 'RefSeq_noversion',\n",
       " 'UniProt_IDs',\n",
       " 'PDB_path',\n",
       " 'Residue_position',\n",
       " 'Residue',\n",
       " 'DSSP_path',\n",
       " 'intra_contacts',\n",
       " 'is_catalytic',\n",
       " 'pLDDT',\n",
       " 'secondary_structure',\n",
       " 'accessibility',\n",
       " 'total energy',\n",
       " 'Backbone Hbond',\n",
       " 'Sidechain Hbond',\n",
       " 'Van der Waals',\n",
       " 'Electrostatics',\n",
       " 'Solvation Polar',\n",
       " 'Solvation Hydrophobic',\n",
       " 'Van der Waals clashes',\n",
       " 'entropy sidechain',\n",
       " 'entropy mainchain',\n",
       " 'sloop_entropy',\n",
       " 'mloop_entropy',\n",
       " 'cis_bond',\n",
       " 'torsional clash',\n",
       " 'backbone clash',\n",
       " 'helix dipole',\n",
       " 'water bridge',\n",
       " 'disulfide',\n",
       " 'electrostatic kon',\n",
       " 'partial covalent bonds',\n",
       " 'energy Ionisation',\n",
       " 'Entropy Complex']"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gnomad_variants.columns.values.tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6bba952d-0e3f-4438-805f-e39df1fb4a5b",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
